---
title: "Cell type homology"
output: html_notebook
---


```{r init-workspace, message=FALSE, warnings=FALSE}
# Load libraries 
library(reshape2)
library(ggplot2)
library(ggrepel)
# library(biomaRt)
library(limma)
library(RColorBrewer)

# Load functions 
titlecase <- function (x) {
  # TODO: use Rex package
  gsub("\\b([a-z])([a-z]+)", "\\U\\1\\L\\2", tolower(x), perl = TRUE)
}

```


##### Which low expressing genes are picked up by different methods?
```{r load-data, message=FALSE}
# 11/2016
# load("//AIBSDATA/humancelltypes/Analyses/CellTypes/Cell_vs_Nuc/data/mouse_v1_cells_nuclei.RData", 
     # verbose = TRUE)

# Exons + Introns
load("//allen/programs/celltypes/workgroups/hct/CT_clustering/mouse_cell_vs_nuc/data/20170818_VISp_L5_nuc_iter_cl_data.rda", verbose = TRUE)
samp.dat.nuc <- samp.dat
nbt.data.nuc <- nbt.data
load("//allen/programs/celltypes/workgroups/hct/CT_clustering/mouse_cell_vs_nuc/data/20170818_VISp_L5_cell_iter_cl_data.rda", verbose = TRUE)
samp.dat.cell <- samp.dat
nbt.data.cell <- nbt.data

# Exons only
load("//allen/programs/celltypes/workgroups/hct/CT_clustering/mouse_cell_vs_nuc/data/20170818_VISp_L5_nuc_exon_iter_cl_data.rda", verbose = TRUE)
samp.dat.nuc.exon <- samp.dat
nbt.data.nuc.exon <- nbt.data
load("//allen/programs/celltypes/workgroups/hct/CT_clustering/mouse_cell_vs_nuc/data/20170818_VISp_L5_cell_exon_iter_cl_data.rda", verbose = TRUE)
samp.dat.cell.exon <- samp.dat
nbt.data.cell.exon <- nbt.data


# Load correlation data
load("../analysis/match_nuc_cell/map_nuc_to_cells.rda")

# Load gene lengths
gene.len <- read.csv(file = "../lib/mouse_GRCm38_gene_len.csv")
gene.len$gene_orig <- gene.len$gene
gene.len$gene <- toupper(gene.len$gene)

```


```{r process-data}
samp.dat <- rbind(samp.dat.nuc, samp.dat.cell)
samp.dat$Samples <- as.factor(samp.dat$cell_prep_type)

nuc.cell.genes <- intersect(row.names(nbt.data.nuc), row.names(nbt.data.cell))
cpm.dat <- cbind(nbt.data.nuc[nuc.cell.genes, ], nbt.data.cell[nuc.cell.genes, ])
rownames(cpm.dat) <- toupper(rownames(cpm.dat))

```

## Cell vs. nuclear gene detection
```{r single-cell-detection, message=FALSE, warning=FALSE}
# Single Rbp4+ cell gene expression detection
keep.samp <- 1:nrow(samp.dat)
samp.subset <- droplevels(samp.dat[keep.samp, ])
cpm.subset <- droplevels(cpm.dat[, keep.samp])

# Single cell gene detection
cpm.subset.mean <- t(apply(cpm.subset, 1, 
                            function(x) tapply(x, samp.subset$Samples, mean)))

genes <- apply(cpm.subset.mean, 2, function(x) rownames(cpm.subset.mean)[x > 1])
cell.genes <- setdiff(genes$Cells, genes$Nuclei)
nuc.genes <- setdiff(genes$Nuclei, genes$Cells)

write.csv(cell.genes, file="../analysis/nuc_vs_cell/cell.genes_gt1.csv", row.names=FALSE)
write.csv(nuc.genes, file="../analysis/nuc_vs_cell/nuc.genes_gt1.csv", row.names=FALSE)

write.csv(genes$Nuclei, file="../analysis/nuc_vs_cell/nuc.all.genes_gt1.csv", row.names=FALSE)
write.csv(genes$Cells, file="../analysis/nuc_vs_cell/cell.all.genes_gt1.csv", row.names=FALSE)


# Plot expression correlation within methods (single cell)
cor.all <- data.frame()
par(mfrow=c(2, 2))
for (i in 1:nlevels(samp.subset$Samples)) {
  method1 <- levels(samp.subset$Samples)[i]
  col.subset <- which(samp.subset$Samples == method1)
  corm <- cor(cpm.subset[, col.subset])
  cor.all <- rbind(cor.all, data.frame(Samples = method1, Cor_method = "Unweighted",
                                       Correlation = corm[upper.tri(corm)]))
  
  # Plot pair with max cor
  diag(corm) <- 0
  samp1 <- row(corm)[which.max(corm)]
  samp2 <- col(corm)[which.max(corm)]
  cor1 <- cor(cpm.subset[, col.subset[samp1]], cpm.subset[, col.subset[samp2]])
  samp.pair <- data.frame(samp1 = cpm.subset[, col.subset[samp1]], 
                          samp2 = cpm.subset[, col.subset[samp2]])
  g1 <- ggplot(samp.pair, aes(x = samp1, y = samp2)) +
    geom_abline(intercept = 0, slope = 1) +
    geom_point(shape = 1) +
    xlim(c(0, 10)) +
    ylim(c(0, 10)) +
    xlab(paste(method1, 1)) +
    ylab(paste(method1, 2)) +
    ggtitle(paste("r =", round(cor1, 2))) +
    theme_bw()
  plot(g1)
  ggsave(g1, width = 3, height = 3, dpi = 600,
         filename = paste0("../analysis/nuc_vs_cell/", method1, "_pair.png"))
}

# Add weighted pairwise cor (previously computed)
cor.all <- rbind(cor.all,
                 data.frame(Samples = "Cells", Cor_method = "Weighted",
                            Correlation = cor.cell[upper.tri(cor.cell)]),
                 data.frame(Samples = "Nuclei", Cor_method = "Weighted",
                            Correlation = cor.nuc.ordered[upper.tri(cor.nuc.ordered)]))

cor.all$Samples <- factor(cor.all$Samples, levels = c("Nuclei", "Cells"))
g0 <- ggplot(cor.all, aes(x = Correlation, color = Samples, linetype = Cor_method)) + 
  geom_density(size = 1) + 
  geom_hline(yintercept = 0, color = "grey80", size = 1) +
  scale_color_brewer(palette = "Set1") +
  ggtitle("Pairwise correlation between samples") +
  theme_bw()
plot(g0)
ggsave(g0, width = 4, height = 3, 
       filename = "../analysis/nuc_vs_cell/nuc_vs_cell_pw_cor.pdf")


# Count # cells expressing each gene
for (split.v4 in c(FALSE, TRUE, FALSE)) {  # 
  if (split.v4 == TRUE) {
    samp.subset$Samples <- as.character(samp.subset$Samples)
    cell.ids <- which(samp.subset$Samples == "Cells")
    replace.names <- sample(cell.ids, round(length(cell.ids) / 2), replace = FALSE)
    samp.subset$Samples[replace.names] <- "Cellsb"
    samp.subset$Samples <- as.factor(samp.subset$Samples)
  } else {
    samp.subset <- droplevels(samp.dat[keep.samp, ])
  }
  
  gene.cnt <- as.data.frame(matrix(NA, nrow(cpm.subset), 
                                   nlevels(samp.subset$Samples)))
  rownames(gene.cnt) <- rownames(cpm.subset)
  colnames(gene.cnt) <- levels(samp.subset$Samples)
  gene.expr <- as.data.frame(matrix(NA, nrow(cpm.subset), 
                                    nlevels(samp.subset$Samples)))
  rownames(gene.expr) <- rownames(cpm.subset)
  colnames(gene.expr) <- levels(samp.subset$Samples)
  for (i in 1:nlevels(samp.subset$Samples)) {
    method1 <- levels(samp.subset$Samples)[i]
    col.subset <- which(samp.subset$Samples == method1)
    cnt1 <- apply(cpm.subset[, col.subset], 1, function(x) sum(x > 0))
    gene.cnt[, i] <- cnt1
    expr1 <- apply(cpm.subset[, col.subset], 1, function(x) mean(x[x > 0]))
    expr1[is.na(expr1)] <- 0
    gene.expr[, i] <- expr1
  }
  
  # Plot proportion of cells where gene is detected
  expr.mean <- apply(gene.expr, 1, mean)
  # pairs(gene.cnt, cex = expr.mean)
  
  # plot(gene.cnt$Nuclei, gene.cnt$Cells, type = "n")
  # text(gene.cnt$Nuclei, gene.cnt$Cells,
  #      rownames(gene.cnt), cex = 1)
  
  gene.prop <- sweep(gene.cnt, 2, apply(gene.cnt, 2, max), "/")
  # write.csv(gene.prop, file = "../analysis/nuc_vs_cell/gene.prop.csv")
  # keep.genes <- order(apply(gene.prop, 1, sd), decreasing = TRUE)[1:1000]
  keep.genes <- abs(gene.prop$Cells - gene.prop$Nuclei) > 0.4
  # pheatmap(gene.prop[keep.genes, ], fontsize_row = 2, fontsize_col = 36, 
  # cutree_rows = 3, clustering_distance_rows = "euclidean")
  
  # Plot expression vs. consistency
  gene.expr.prop <- data.frame(gene = rownames(gene.expr),
                               Cell_expr = gene.expr$Cells, 
                               Nuc_expr = gene.expr$Nuclei, 
                               Nucleus_prop = gene.prop$Nuclei,
                               Cell_prop = gene.prop$Cells,
                               Cell_Nuc_diff = gene.prop$Cells - gene.prop$Nuclei)
  rownames(gene.expr.prop) <- rownames(gene.expr)
  
  if (split.v4 == TRUE) {
    cut1 <- cut(gene.expr.prop$Cell_expr, c(0, seq(0.01, 12, by=0.2)), include.lowest = TRUE)
    prop.diff <- abs(gene.prop$Cells - gene.prop$Cellsb)
    prop.exp <- data.frame(Cell_expr = c(0, seq(0.11, 11.9, 0.2)), 
                           exprq = tapply(prop.diff, cut1, quantile, 0.975))
    cut2 <- cut(gene.prop$Cells, seq(0, 1, by=0.02), include.lowest = TRUE)
    prop.exp2 <- data.frame(Cell_expr = seq(0.01, 0.99, 0.02), 
                           exprq = tapply(gene.prop$Cellsb, cut2, quantile, 0.975))
  }
}

# Save list of genes
out.fn1 <- "../analysis/nuc_vs_cell/Cell_vs_nuc_gene_detection.csv"
if (! file.exists(out.fn1)) {
  write.csv(gene.expr.prop, file = out.fn1, row.names = FALSE)
}
```



```{r dex-genes, message = FALSE, warning = FALSE}
# DEX genes
design <- model.matrix(~ 0 + samp.subset$Samples)
colnames(design) <- gsub("samp.subset$Samples", "", colnames(design), fixed = TRUE)
# design <- design[sample(1:nrow(design), nrow(design), replace = FALSE), ]  # Perm control

fit <- lmFit(cpm.subset, design)
cont.matrix <- makeContrasts(contrasts="Cells-Nuclei", levels = design)

fit2 <- eBayes(contrasts.fit(fit, cont.matrix))
dex.ie <- topTable(fit2, number = Inf, p.value = 1, adjust = "BH", sort.by = "none")
dex.ie$gene <- row.names(dex.ie)
# dex.ie <- subset(dex.ie, AveExpr > 0)

out.fn2 <- "../analysis/nuc_vs_cell/Cell_vs_nuc_dex_genes_introns_exons.csv"
if (! file.exists(out.fn2)) {
  write.csv(dex.ie, file = out.fn2, row.names = FALSE)
}
```



## Cell differential detection vs. DEX
```{r gene-prop-vs-dex, message = FALSE, warning = FALSE, fig.width = 5, fig.height = 4}
genes.expr <- apply(gene.expr.prop, 1, function(x) ! all(x == 0))
gene.prop.dex <- merge(gene.expr.prop[genes.expr, ], dex.ie, by = "gene")
colnames(gene.prop.dex) <- sub("Row.names", "gene", colnames(gene.prop.dex),
                               fixed = TRUE)
gene.annot <- subset(gene.prop.dex, gene %in% 
                       c("FOS", "ARC", "EGR1", "RBFOX3", "RBP4"))

pal.spectral <- colorRampPalette(rev(brewer.pal(11,'Spectral')))(100)

# Prop nuc vs. prop Cells
g5 <- ggplot(gene.prop.dex, aes(x = Nucleus_prop, y = Cell_prop)) +
  geom_hline(yintercept = 0) +
  stat_bin_hex(bins = 50) +
  scale_fill_gradientn(colours=pal.spectral, trans="log10", name="No. of genes") +
  xlab("Nuclei gene detection") + 
  ylab("Cells gene detection") + 
  geom_abline(intercept = 0, slope = 1, color = "grey") +
  geom_smooth(data = prop.exp2, aes(Cell_expr, exprq), 
              color = "grey", size = 1, se = FALSE) +
  geom_smooth(data = prop.exp2, aes(exprq, Cell_expr), 
              color = "grey", size = 1, se = FALSE) +
  theme_bw() + 
  theme(panel.grid.minor = element_blank(),
        axis.title = element_text(size=12))
plot(g5)
ggsave(g5, width = 4.25, height = 3, 
       filename = "../analysis/nuc_vs_cell/nuc_vs_cell_prop_scatter.pdf")


# DEX vs. sig
g2 <- ggplot(gene.prop.dex, aes(x = logFC, y = -log10(adj.P.Val))) +
  stat_bin_hex(bins = 50) +
  geom_hline(yintercept = -log10(0.05), color = "grey", size = 0.5) +
  geom_vline(xintercept = c(-log2(1.5), log2(1.5)), color = "grey", size = 0.5) +
  xlim(c(-6, 6)) +
  scale_fill_gradientn(colours=pal.spectral, trans="log10", name="No. of genes") +
  xlab("Cells vs. Nuclei (log2 fold change)") +
  ylab("Significance (-log10 P-value)") +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())
plot(g2)
ggsave(g2, width = 4.25, height = 3, 
       filename = "../analysis/nuc_vs_cell/nuc_vs_cell_dex_volcano.pdf")


# Prop vs. expr
g4 <- ggplot(gene.prop.dex, aes(Cell_expr, Cell_Nuc_diff)) +
  geom_hline(yintercept = 0) +
  stat_bin_hex(bins = 50) +
  scale_fill_gradientn(colours=pal.spectral, trans="log10", name="# Genes") +
  geom_smooth(data = prop.exp, aes(Cell_expr, exprq), 
              color = "grey", size = 2, se = FALSE) +
  geom_smooth(data = prop.exp, aes(Cell_expr, -exprq), 
              color = "grey", size = 2, se = FALSE) +
  # geom_text(data = gene.annot, aes(x = logFC, y = Cell_Nuc_diff, label = gene),
  #           fontface = 3, size = 3) +
  xlab("Cell mean expression (cells with RPKM > 0)") + 
  ylab("Difference in proportion of cells expressing gene\n(Cell - Nucleus)") + 
  ylim(c(-1, 1)) +
  theme_bw() + 
  theme(axis.title = element_text(size=12))
# plot(g4)


# DEX vs. Prop
g3 <- ggplot(gene.prop.dex, aes(x = logFC, y = Cell_Nuc_diff)) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  stat_bin_hex(bins = 50) +
  scale_fill_gradientn(colours=pal.spectral, trans="log10") +
  geom_text(data = gene.annot, aes(x = logFC, y = Cell_Nuc_diff, label = gene), 
            fontface = 3, size = 3) +
  xlab("Cell vs. Nucleus expression (log2 fold change)") +
  ylab("Difference in proportion of cells expressing gene\n(Cell - Nucleus)") + 
  theme_bw()
# plot(g3)

# DEX vs. expr
g1 <- ggplot(gene.prop.dex, aes(x = Cell_expr, y = logFC)) +
  geom_hline(yintercept = 0) +
  stat_bin_hex(bins = 50) +
  scale_fill_gradientn(colours=pal.spectral, trans="log10") +
  xlab("Cell mean expression (cells with RPKM > 0)") +
  ylab("Cell vs. Nucleus expression (log2 fold change)") +
  theme_bw()
# plot(g1)

```



```{r gene-distrib-cellvsnuc, fig.width = 4.25, fig.height = 4}
nuc.enriched.genes <- as.character(gene.prop.dex$gene[order(gene.prop.dex$logFC,
                                                           decreasing = FALSE)])
cell.enriched.genes <- as.character(gene.prop.dex$gene[order(gene.prop.dex$logFC,
                                                           decreasing = TRUE)])

plot.genes <- list()
plot.genes[["nuc_vs_wc"]] <- c("CACNA1C", "KCND2","KCNH7", # ion ch  (CACNA1D)
                               "GRIK2", "GRIK3", "GRM8",  # glut signal
                               "CNTNAP2","ROBO2","SEMA6D")  # axon guidance (CNTN4)
plot.genes[["wc_vs_nuc"]] <- c(cell.enriched.genes[grep("^PSM", cell.enriched.genes)][1],  # proteasome
                               cell.enriched.genes[grep("^RPL", cell.enriched.genes)][1:2],  # ribosome
                               c("POLR2J", "SNRPD1", "UBL5"),  # RNA processing
                               c("FOS", "EGR1", "ARC"))  # IEG
# plot.genes[["K_ch"]] <- nuc.enriched.genes[grep("^KCN", nuc.enriched.genes)[1:9]]
# plot.genes[["Ca_ch"]] <- nuc.enriched.genes[grep("^CACNA", nuc.enriched.genes)[1:9]]
plot.genes <- lapply(plot.genes, titlecase)

all.plot.genes <- unique(unlist(plot.genes))
rpkmw <- cpm.subset[toupper(all.plot.genes), ]
rpkmw$gene <- titlecase(rownames(rpkmw))
rpkmw$gene <- factor(rpkmw$gene, levels = all.plot.genes)
rpkml <- melt(rpkmw, id.vars = "gene")
rpkml$cellnuc <- rep(samp.subset$Samples, each = length(rpkmw$gene))
# rpkml$cellnuc <- factor(rpkml$cellnuc, levels = c("Nuclei", "Cells"))

expr.violins <- list()
for (gene.list in names(plot.genes)) {
  g1 <- ggplot(subset(rpkml, gene %in% plot.genes[[gene.list]]), 
               aes(x = cellnuc, y = value, fill = cellnuc)) + 
    facet_wrap(~ gene, nrow = 3, dir = "v") +
    geom_violin(show.legend = FALSE) +
    scale_fill_manual(values = brewer.pal(3, "Set1")[2:1]) +
    xlab("") +
    ylab("Expression (log2 CPM + 1)") +
    ggtitle(gene.list) +
    theme_bw() +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          strip.text = element_text(face = "italic"))
  plot(g1)
  expr.violins[[gene.list]] <- g1
}


ggsave(expr.violins[["nuc_vs_wc"]], width = 3.25, height = 4,
       filename = paste0("../analysis/nuc_vs_cell/nuc_vs_wc_violin.pdf"))

ggsave(expr.violins[["wc_vs_nuc"]], width = 3.25, height = 4,
       filename = paste0("../analysis/nuc_vs_cell/wc_vs_nuc_violin.pdf"))



```


```{r gene-length-vs-nuc-enrichment, fig.width = 5, fig.height = 3, warning=FALSE}
gene.prop.dex2 <- merge(gene.prop.dex, gene.len, by = "gene")
gene.prop.dex2$logFC_bin <- cut(gene.prop.dex2$logFC, breaks = seq(-2, 5.5, 0.5))
gene.prop.dex2 <- gene.prop.dex2[order(gene.prop.dex2$AveExpr), ]

nuc.short.genes <- subset(gene.prop.dex2, logFC < -log2(1.5) & gene_len < 1e5 & AveExpr > 10)
cell.long.genes <- subset(gene.prop.dex2, logFC > 0.5 & AveExpr > 7 &
                            gene_len > 1e6)

# write.csv(nuc.short.genes$gene, file = "../genes.csv")

g1 <- ggplot(gene.prop.dex2, aes(x = logFC, y = gene_len, color = AveExpr)) +
  # geom_vline(xintercept = 0, size = 1, color = "grey") +
  geom_point() +
  # geom_text_repel(data=nuc.short.genes, aes(label = gene), color = "black") +
  # geom_vline(xintercept = c(-log2(1.5), log2(1.5)), size = 1, color = "grey") +
  # geom_text(data=cell.long.genes, aes(label = gene), 
  #           color = "black", hjust = 0, nudge_x = 0.1) +
  xlab("Nuclei vs. Cells (log2 fold change)") +
  ylab("Genomic length (base pairs)") +
  scale_color_gradientn(colors = grey.colors(10, 0.9, 0, gamma = 1)) +
  theme_bw()
plot(g1)
ggsave(g1, width = 5, height = 3, dpi = 600,
         filename = paste0("../analysis/nuc_vs_cell/gene_length_vs_nuc_enrichment.png"))


```


```{r exon-only-dex-genes}
nuc.cell.exon.genes <- intersect(row.names(nbt.data.nuc.exon), 
                                 row.names(nbt.data.cell.exon))
cpm.dat.exon <- cbind(nbt.data.nuc.exon[nuc.cell.exon.genes, ], 
                      nbt.data.cell.exon[nuc.cell.exon.genes, ])
rownames(cpm.dat.exon) <- toupper(rownames(cpm.dat.exon))


# DEX genes
design <- model.matrix(~ 0 + samp.subset$Samples)
colnames(design) <- gsub("samp.subset$Samples", "", colnames(design), fixed = TRUE)

fit <- lmFit(cpm.dat.exon, design)
cont.matrix <- makeContrasts(contrasts="Cells-Nuclei", levels = design)

fit2 <- eBayes(contrasts.fit(fit, cont.matrix))
dex.exons <- topTable(fit2, number = Inf, p.value = 1, adjust = "BH", sort.by = "none")
# dex.exons <- subset(dex.exons, AveExpr > 0)
dex.exons$gene <- row.names(dex.exons)


out.fn3 <- "../analysis/nuc_vs_cell/Cell_vs_nuc_dex_genes_exons_only.csv"
if (! file.exists(out.fn3)) {
  write.csv(dex.exons, file = out.fn3, row.names = FALSE)
}


```


```{r compare-dex-byreads, fig.width = 8, fig.height=4}
gene.prop.dex3 <- merge(gene.prop.dex2, dex.exons, by = "gene", 
                        suffixes = c("_ie", "_ex"))

gdex <- ggplot(gene.prop.dex3, aes(x = logFC_ie, y = logFC_ex)) +
  geom_abline(intercept = 0, slope = 1) +
  # geom_text_repel() +
    # geom_hline(yintercept = c(-log2(1.5), log2(1.5))) +
  # geom_vline(xintercept = c(-log2(1.5), log2(1.5))) +
  geom_point() +
  theme_bw()
plot(gdex)

subset(gene.prop.dex3, logFC_ie > log2(1.5) & adj.P.Val_ex > 0.05)

subset(gene.prop.dex3, logFC_ie < -log2(1.5) & adj.P.Val_ex > 0.05)
subset(gene.prop.dex3, logFC_ex < -log2(1.5) & adj.P.Val_ie > 0.05)

subset(gene.prop.dex3, logFC_ie < -log2(1.5) & logFC_ex < -log2(1.5) &
         adj.P.Val_ie < 0.05 & adj.P.Val_ex < 0.05)

subset(gene.prop.dex3, (logFC_ie < -log2(1.5) & adj.P.Val_ie < 0.05) &
                        (logFC_ex < -log2(1.5) & adj.P.Val_ex < 0.05))




pdf(file = paste0("../analysis/nuc_vs_cell/gene_length_ecdf.pdf"),
    width = 8, height = 4)

par(mfrow = c(1, 2))
plot.ecdf(gene.prop.dex3$gene_len[gene.prop.dex3$AveExpr_ie > 0],
          xlim = c(1e1, 1e7), log = "x", las = 1,
          xlab = "Length (base pairs)", ylab = "Cumulative proportion",
          main = "Genomic length")
lines(ecdf(gene.prop.dex3$gene_len[gene.prop.dex3$logFC_ie > log2(1.5)]), 
          col = "blue", verticals=T, do.points=F,col.01line = NULL)
lines(ecdf(gene.prop.dex3$gene_len[gene.prop.dex3$logFC_ie < -log2(1.5)]), 
      col = "red", verticals=T, do.points=F,col.01line = NULL)
lines(ecdf(gene.prop.dex3$gene_len[gene.prop.dex3$logFC_ex > log2(1.5)]), 
      col = "light blue", verticals=T, do.points=F,col.01line = NULL)
lines(ecdf(gene.prop.dex3$gene_len[gene.prop.dex3$logFC_ex < -log2(1.5)]), 
      col = "pink", verticals=T, do.points=F,col.01line = NULL)


plot.ecdf(gene.prop.dex3$transcript_len[gene.prop.dex3$AveExpr_ie > 0], 
          xlim = c(1e1, 1e7), log = "x", las = 1,
          xlab = "Length (base pairs)", ylab = "Cumulative proportion",
          main = "Transcript length")
lines(ecdf(gene.prop.dex3$transcript_len[gene.prop.dex3$logFC_ie > log2(1.5)]), 
          col = "blue", verticals=T, do.points=F,col.01line = NULL)
lines(ecdf(gene.prop.dex3$transcript_len[gene.prop.dex3$logFC_ie < -log2(1.5)]), 
      col = "red", verticals=T, do.points=F,col.01line = NULL)
lines(ecdf(gene.prop.dex3$transcript_len[gene.prop.dex3$logFC_ex > log2(1.5)]), 
      col = "light blue", verticals=T, do.points=F,col.01line = NULL)
lines(ecdf(gene.prop.dex3$transcript_len[gene.prop.dex3$logFC_ex < -log2(1.5)]), 
      col = "pink", verticals=T, do.points=F,col.01line = NULL)

legend("bottomright", bty = "n", cex = 0.5, 
       fill = c("red", "pink", "blue", "light blue", "black"),
       legend = c("Nuclei (introns + exons)",
                  "Nuclei (exons)",
                  "Cells (introns + exons)",
                  "Cells (exons)",
                  "All expressed genes"))

dev.off()



```


```{r go-treemaps}
revigo.data.list <- list()

#### Cell > 4*Nuc (introns + exons) ####
revigo.data.list[["CC_cell_ie"]] <- rbind(c("GO:0005925","focal adhesion",0.109,12.7773,0.968,0.000,"focal adhesion"),
c("GO:0022626","cytosolic ribosome",0.187,38.2480,0.522,0.000,"cytosolic ribosome"),
c("GO:0005694","chromosome",1.505,5.0625,0.646,0.633,"cytosolic ribosome"),
c("GO:0016272","prefoldin complex",0.056,3.5031,0.714,0.361,"cytosolic ribosome"),
c("GO:0005697","telomerase holoenzyme complex",0.025,1.9830,0.534,0.505,"cytosolic ribosome"),
c("GO:0090575","RNA polymerase II transcription factor complex",0.155,2.1349,0.470,0.585,"cytosolic ribosome"),
c("GO:0008023","transcription elongation factor complex",0.080,1.6253,0.473,0.662,"cytosolic ribosome"),
c("GO:0033290","eukaryotic 48S preinitiation complex",0.089,9.5331,0.593,0.419,"cytosolic ribosome"),
c("GO:0016281","eukaryotic translation initiation factor 4F complex",0.002,1.7852,0.664,0.296,"cytosolic ribosome"),
c("GO:0089701","U2AF",0.017,1.7375,0.534,0.491,"cytosolic ribosome"),
c("GO:0005730","nucleolus",0.664,33.6144,0.507,0.413,"cytosolic ribosome"),
c("GO:0005732","small nucleolar ribonucleoprotein complex",0.072,5.0625,0.658,0.412,"cytosolic ribosome"),
c("GO:0036513","Derlin-1 retrotranslocation complex",0.005,2.9508,0.546,0.656,"cytosolic ribosome"),
c("GO:0005635","nuclear envelope",0.283,2.4157,0.494,0.617,"cytosolic ribosome"),
c("GO:1990131","Gtr1-Gtr2 GTPase complex",0.001,1.4283,0.773,0.186,"cytosolic ribosome"),
c("GO:0005681","spliceosomal complex",0.250,18.6990,0.468,0.610,"cytosolic ribosome"),
c("GO:0000803","sex chromosome",0.007,1.4789,0.730,0.408,"cytosolic ribosome"),
c("GO:0097346","INO80-type complex",0.056,2.9393,0.497,0.538,"cytosolic ribosome"),
c("GO:0036019","endolysosome",0.001,2.3036,0.612,0.612,"cytosolic ribosome"),
c("GO:0036020","endolysosome membrane",0.000,1.8794,0.578,0.643,"cytosolic ribosome"),
c("GO:0000805","X chromosome",0.002,3.0137,0.745,0.273,"cytosolic ribosome"),
c("GO:0005834","heterotrimeric G-protein complex",0.046,1.8125,0.589,0.530,"cytosolic ribosome"),
c("GO:0005840","ribosome",4.198,22.7520,0.425,0.589,"cytosolic ribosome"),
c("GO:0005844","polysome",0.037,3.0013,0.672,0.392,"cytosolic ribosome"),
c("GO:0000407","pre-autophagosomal structure",0.033,3.1612,0.752,0.253,"cytosolic ribosome"),
c("GO:0036464","cytoplasmic ribonucleoprotein granule",0.127,6.3686,0.544,0.430,"cytosolic ribosome"),
c("GO:0000974","Prp19 complex",0.017,2.1891,0.733,0.333,"cytosolic ribosome"),
c("GO:0030904","retromer complex",0.032,3.9431,0.592,0.516,"cytosolic ribosome"),
c("GO:1903293","phosphatase complex",0.073,6.3969,0.689,0.545,"cytosolic ribosome"),
c("GO:0030906","retromer, cargo-selective complex",0.002,2.6021,0.641,0.434,"cytosolic ribosome"),
c("GO:0097422","tubular endosome",0.002,2.0969,0.582,0.580,"cytosolic ribosome"),
c("GO:0008180","COP9 signalosome",0.031,2.7423,0.519,0.513,"cytosolic ribosome"),
c("GO:0005850","eukaryotic translation initiation factor 2 complex",0.006,1.4283,0.649,0.311,"cytosolic ribosome"),
c("GO:0005851","eukaryotic translation initiation factor 2B complex",0.009,1.7375,0.642,0.319,"cytosolic ribosome"),
c("GO:0099568","cytoplasmic region",0.265,2.2104,0.717,0.307,"cytosolic ribosome"),
c("GO:0005852","eukaryotic translation initiation factor 3 complex",0.117,8.7932,0.586,0.277,"cytosolic ribosome"),
c("GO:0005853","eukaryotic translation elongation factor 1 complex",0.012,2.6021,0.636,0.325,"cytosolic ribosome"),
c("GO:0043596","nuclear replication fork",0.061,1.5421,0.557,0.695,"cytosolic ribosome"),
c("GO:0031461","cullin-RING ubiquitin ligase complex",0.159,6.7212,0.578,0.695,"cytosolic ribosome"),
c("GO:0015934","large ribosomal subunit",0.559,17.8447,0.468,0.666,"cytosolic ribosome"),
c("GO:0005768","endosome",0.319,12.2000,0.507,0.312,"cytosolic ribosome"),
c("GO:0005774","vacuolar membrane",0.290,15.4597,0.508,0.298,"cytosolic ribosome"),
c("GO:0005773","vacuole",0.455,11.6819,0.627,0.324,"cytosolic ribosome"),
c("GO:0070822","Sin3-type complex",0.010,1.6778,0.496,0.519,"cytosolic ribosome"),
c("GO:0031984","organelle subcompartment",0.269,2.4724,0.618,0.374,"cytosolic ribosome"),
c("GO:0048500","signal recognition particle",0.100,2.8013,0.594,0.422,"cytosolic ribosome"),
c("GO:0005785","signal recognition particle receptor complex",0.011,1.4283,0.509,0.690,"cytosolic ribosome"),
c("GO:0005786","signal recognition particle, endoplasmic reticulum targeting",0.034,2.8013,0.619,0.390,"cytosolic ribosome"),
c("GO:0048471","perinuclear region of cytoplasm",0.135,5.6799,0.730,0.280,"cytosolic ribosome"),
c("GO:0005793","endoplasmic reticulum-Golgi intermediate compartment",0.026,4.6459,0.688,0.249,"cytosolic ribosome"),
c("GO:0005789","endoplasmic reticulum membrane",0.761,8.9547,0.486,0.653,"cytosolic ribosome"),
c("GO:0048770","pigment granule",0.016,4.2907,0.613,0.674,"cytosolic ribosome"),
c("GO:0001650","fibrillar center",0.035,1.7852,0.581,0.690,"cytosolic ribosome"),
c("GO:0070993","translation preinitiation complex",0.093,8.7932,0.596,0.420,"cytosolic ribosome"),
c("GO:0097525","spliceosomal snRNP complex",0.128,9.9914,0.472,0.575,"cytosolic ribosome"),
c("GO:0034708","methyltransferase complex",0.080,6.6383,0.602,0.659,"cytosolic ribosome"),
c("GO:0072669","tRNA-splicing ligase complex",0.006,2.0381,0.698,0.312,"cytosolic ribosome"),
c("GO:0044452","nucleolar part",0.153,10.1911,0.547,0.642,"cytosolic ribosome"),
c("GO:0044445","cytosolic part",0.496,33.2403,0.695,0.617,"cytosolic ribosome"),
c("GO:0035770","ribonucleoprotein granule",0.131,6.0706,0.576,0.431,"cytosolic ribosome"),
c("GO:1990904","ribonucleoprotein complex",5.291,59.3706,0.669,0.543,"cytosolic ribosome"),
c("GO:0070971","endoplasmic reticulum exit site",0.006,2.9508,0.614,0.447,"cytosolic ribosome"),
c("GO:0072686","mitotic spindle",0.078,1.3979,0.622,0.693,"cytosolic ribosome"),
c("GO:0044427","chromosomal part",1.117,4.0357,0.548,0.608,"cytosolic ribosome"),
c("GO:0071541","eukaryotic translation initiation factor 3 complex, eIF3m",0.011,5.7399,0.637,0.324,"cytosolic ribosome"),
c("GO:0002199","zona pellucida receptor complex",0.003,2.9508,0.757,0.297,"cytosolic ribosome"),
c("GO:0071013","catalytic step 2 spliceosome",0.062,14.1397,0.466,0.541,"cytosolic ribosome"),
c("GO:0030117","membrane coat",0.264,6.7905,0.535,0.405,"cytosolic ribosome"),
c("GO:0034709","methylosome",0.006,8.9208,0.601,0.447,"cytosolic ribosome"),
c("GO:0034715","pICln-Sm protein complex",0.006,4.1355,0.649,0.312,"cytosolic ribosome"),
c("GO:0000502","proteasome complex",0.389,19.1752,0.540,0.636,"cytosolic ribosome"),
c("GO:0034719","SMN-Sm protein complex",0.008,3.8327,0.643,0.318,"cytosolic ribosome"),
c("GO:0008287","protein serine/threonine phosphatase complex",0.073,6.3969,0.652,0.545,"cytosolic ribosome"),
c("GO:0030532","small nuclear ribonucleoprotein complex",0.138,9.1669,0.487,0.579,"cytosolic ribosome"),
c("GO:0016592","mediator complex",0.155,3.7447,0.453,0.694,"cytosolic ribosome"),
c("GO:0033116","endoplasmic reticulum-Golgi intermediate compartment membrane",0.008,3.8827,0.626,0.574,"cytosolic ribosome"),
c("GO:0032588","trans-Golgi network membrane",0.008,1.3188,0.578,0.573,"cytosolic ribosome"),
c("GO:0034045","pre-autophagosomal structure membrane",0.018,1.8356,0.712,0.646,"cytosolic ribosome"),
c("GO:0016607","nuclear speck",0.091,4.3206,0.559,0.614,"cytosolic ribosome"),
c("GO:0033176","proton-transporting V-type ATPase complex",0.146,2.5768,0.675,0.579,"cytosolic ribosome"),
c("GO:0015630","microtubule cytoskeleton",0.900,10.3298,0.640,0.592,"cytosolic ribosome"),
c("GO:0030687","preribosome, large subunit precursor",0.067,4.2708,0.655,0.410,"cytosolic ribosome"),
c("GO:0033202","DNA helicase complex",0.093,2.0004,0.628,0.556,"cytosolic ribosome"),
c("GO:0030684","preribosome",0.223,6.8386,0.633,0.600,"cytosolic ribosome"),
c("GO:0070603","SWI/SNF superfamily-type complex",0.112,3.6882,0.484,0.569,"cytosolic ribosome"),
c("GO:0031201","SNARE complex",0.038,5.0768,0.597,0.523,"cytosolic ribosome"),
c("GO:0005885","Arp2/3 protein complex",0.063,5.0975,0.530,0.680,"cytosolic ribosome"),
c("GO:0000151","ubiquitin ligase complex",0.232,8.5086,0.575,0.604,"cytosolic ribosome"),
c("GO:0098796","membrane protein complex",2.473,5.8013,0.626,0.533,"cytosolic ribosome"),
c("GO:0030008","TRAPP complex",0.042,1.4461,0.664,0.354,"cytosolic ribosome"),
c("GO:0097255","R2TP complex",0.003,2.0969,0.756,0.299,"cytosolic ribosome"),
c("GO:1990234","transferase complex",1.223,26.7258,0.626,0.439,"cytosolic ribosome"),
c("GO:0042470","melanosome",0.016,4.2907,0.613,0.673,"cytosolic ribosome"),
c("GO:0035145","exon-exon junction complex",0.010,1.9830,0.546,0.473,"cytosolic ribosome"),
c("GO:1902493","acetyltransferase complex",0.152,3.8827,0.643,0.693,"cytosolic ribosome"),
c("GO:1902494","catalytic complex",3.734,27.8633,0.678,0.353,"cytosolic ribosome"),
c("GO:0005952","cAMP-dependent protein kinase complex",0.014,2.0381,0.631,0.476,"cytosolic ribosome"),
c("GO:0030054","cell junction",0.445,3.7235,0.993,0.000,"cell junction"),
c("GO:0045202","synapse",0.299,2.0969,0.993,0.000,"synapse"),
c("GO:0048475","coated membrane",0.264,6.7905,0.984,0.000,"coated membrane"),
c("GO:0030496","midbody",0.040,3.4001,0.921,0.044,"midbody"),
c("GO:0043209","myelin sheath",0.049,10.1701,0.920,0.044,"myelin sheath"),
c("GO:0031252","cell leading edge",0.086,3.7190,0.917,0.046,"cell leading edge"),
c("GO:0044297","cell body",0.087,3.7986,0.917,0.046,"cell body"),
c("GO:0030427","site of polarized growth",0.091,3.0137,0.917,0.046,"site of polarized growth"),
c("GO:0043005","neuron projection",0.190,10.4413,0.825,0.049,"neuron projection"),
c("GO:0032838","cell projection cytoplasm",0.014,2.9788,0.722,0.595,"neuron projection"),
c("GO:0001726","ruffle",0.035,2.2692,0.875,0.634,"neuron projection"),
c("GO:0097458","neuron part",0.320,9.3915,0.910,0.051,"neuron part"),
c("GO:0016234","inclusion body",0.019,2.0947,0.856,0.090,"inclusion body"));


revigo.data.list[["BP_cell_ie"]] <- rbind(c("GO:0002478","antigen processing and presentation of exogenous peptide antigen",0.004,12.4486,0.931,0.000,"antigen processing and presentation of exogenous peptide antigen"),
c("GO:0006289","nucleotide-excision repair",0.310,6.0031,0.796,0.565,"antigen processing and presentation of exogenous peptide antigen"),
c("GO:0042059","negative regulation of epidermal growth factor receptor signaling pathway",0.007,5.9914,0.783,0.559,"antigen processing and presentation of exogenous peptide antigen"),
c("GO:0019882","antigen processing and presentation",0.027,8.5575,0.944,0.561,"antigen processing and presentation of exogenous peptide antigen"),
c("GO:0032107","regulation of response to nutrient levels",0.011,7.7645,0.832,0.576,"antigen processing and presentation of exogenous peptide antigen"),
c("GO:0038093","Fc receptor signaling pathway",0.005,10.1911,0.771,0.505,"antigen processing and presentation of exogenous peptide antigen"),
c("GO:0080135","regulation of cellular response to stress",0.182,9.7670,0.761,0.548,"antigen processing and presentation of exogenous peptide antigen"),
c("GO:0015031","protein transport",2.251,39.7545,0.832,0.000,"protein transport"),
c("GO:0070199","establishment of protein localization to chromosome",0.006,3.4597,0.843,0.537,"protein transport"),
c("GO:0071166","ribonucleoprotein complex localization",0.097,4.6615,0.841,0.575,"protein transport"),
c("GO:0016192","vesicle-mediated transport",1.085,12.6402,0.907,0.344,"protein transport"),
c("GO:0016197","endosomal transport",0.131,13.7399,0.920,0.274,"protein transport"),
c("GO:1903827","regulation of cellular protein localization",0.132,7.1518,0.700,0.680,"protein transport"),
c("GO:0050657","nucleic acid transport",0.100,3.1296,0.905,0.533,"protein transport"),
c("GO:1904874","positive regulation of telomerase RNA localization to Cajal body",0.004,5.0680,0.738,0.484,"protein transport"),
c("GO:0007034","vacuolar transport",0.133,13.5421,0.920,0.275,"protein transport"),
c("GO:0007041","lysosomal transport",0.017,3.1972,0.926,0.229,"protein transport"),
c("GO:0006403","RNA localization",0.118,5.9957,0.871,0.643,"protein transport"),
c("GO:1990173","protein localization to nucleoplasm",0.002,3.2076,0.850,0.482,"protein transport"),
c("GO:0090672","telomerase RNA localization",0.004,4.6440,0.883,0.664,"protein transport"),
c("GO:0090670","RNA localization to Cajal body",0.004,4.6440,0.883,0.662,"protein transport"),
c("GO:0051656","establishment of organelle localization",0.180,7.7011,0.822,0.281,"protein transport"),
c("GO:0051668","localization within membrane",0.023,4.1938,0.854,0.519,"protein transport"),
c("GO:0006913","nucleocytoplasmic transport",0.237,7.6108,0.821,0.617,"protein transport"),
c("GO:0051640","organelle localization",0.223,7.0665,0.831,0.627,"protein transport"),
c("GO:0016482","cytosolic transport",0.076,6.5800,0.838,0.619,"protein transport"),
c("GO:0051169","nuclear transport",0.239,8.0670,0.825,0.677,"protein transport"),
c("GO:0016071","mRNA metabolic process",0.798,50.2774,0.847,0.000,"mRNA metabolism"),
c("GO:0045862","positive regulation of proteolysis",0.078,7.6108,0.668,0.609,"mRNA metabolism"),
c("GO:0061418","regulation of transcription from RNA polymerase II promoter in response to hypoxia",0.003,4.2815,0.753,0.476,"mRNA metabolism"),
c("GO:0031329","regulation of cellular catabolic process",0.093,18.5607,0.705,0.583,"mRNA metabolism"),
c("GO:0018205","peptidyl-lysine modification",0.355,3.1965,0.826,0.452,"mRNA metabolism"),
c("GO:0006369","termination of RNA polymerase II transcription",0.015,3.9957,0.867,0.526,"mRNA metabolism"),
c("GO:0006367","transcription initiation from RNA polymerase II promoter",0.109,4.0414,0.850,0.601,"mRNA metabolism"),
c("GO:0006368","transcription elongation from RNA polymerase II promoter",0.082,9.5638,0.852,0.300,"mRNA metabolism"),
c("GO:0016072","rRNA metabolic process",0.965,38.4179,0.832,0.382,"mRNA metabolism"),
c("GO:0016073","snRNA metabolic process",0.035,6.2204,0.868,0.518,"mRNA metabolism"),
c("GO:0070647","protein modification by small protein conjugation or removal",0.821,20.6440,0.818,0.331,"mRNA metabolism"),
c("GO:0006354","DNA-templated transcription, elongation",0.202,10.1203,0.850,0.280,"mRNA metabolism"),
c("GO:0006353","DNA-templated transcription, termination",0.185,6.4685,0.851,0.320,"mRNA metabolism"),
c("GO:0006352","DNA-templated transcription, initiation",0.766,5.3478,0.834,0.361,"mRNA metabolism"),
c("GO:0044265","cellular macromolecule catabolic process",1.268,44.0757,0.756,0.153,"mRNA metabolism"),
c("GO:0000338","protein deneddylation",0.017,5.2874,0.819,0.683,"mRNA metabolism"),
c("GO:0010608","posttranscriptional regulation of gene expression",0.719,23.2480,0.762,0.256,"mRNA metabolism"),
c("GO:1901566","organonitrogen compound biosynthetic process",14.064,13.0575,0.903,0.391,"mRNA metabolism"),
c("GO:0043603","cellular amide metabolic process",6.879,13.9830,0.890,0.189,"mRNA metabolism"),
c("GO:1903311","regulation of mRNA metabolic process",0.044,3.6576,0.745,0.299,"mRNA metabolism"),
c("GO:0000470","maturation of LSU-rRNA",0.047,4.5591,0.719,0.530,"mRNA metabolism"),
c("GO:0043543","protein acylation",0.202,4.1090,0.836,0.429,"mRNA metabolism"),
c("GO:0006259","DNA metabolic process",5.607,3.3526,0.833,0.383,"mRNA metabolism"),
c("GO:0050684","regulation of mRNA processing",0.035,3.2211,0.732,0.611,"mRNA metabolism"),
c("GO:0031400","negative regulation of protein modification process",0.151,12.3625,0.679,0.418,"mRNA metabolism"),
c("GO:0051439","regulation of ubiquitin-protein ligase activity involved in mitotic cell cycle",0.001,16.8996,0.683,0.611,"mRNA metabolism"),
c("GO:0006521","regulation of cellular amino acid metabolic process",0.018,12.3862,0.796,0.276,"mRNA metabolism"),
c("GO:0002183","cytoplasmic translational initiation",0.025,3.3420,0.836,0.399,"mRNA metabolism"),
c("GO:0043488","regulation of mRNA stability",0.018,12.7011,0.801,0.293,"mRNA metabolism"),
c("GO:0002181","cytoplasmic translation",0.064,10.4750,0.829,0.429,"mRNA metabolism"),
c("GO:0043487","regulation of RNA stability",0.021,12.1029,0.799,0.573,"mRNA metabolism"),
c("GO:0036260","RNA capping",0.112,4.0137,0.839,0.542,"mRNA metabolism"),
c("GO:0016567","protein ubiquitination",0.523,18.9626,0.783,0.470,"mRNA metabolism"),
c("GO:0006412","translation",5.686,25.1959,0.754,0.661,"mRNA metabolism"),
c("GO:0009894","regulation of catabolic process",0.146,18.8894,0.760,0.530,"mRNA metabolism"),
c("GO:0006413","translational initiation",0.518,49.8069,0.801,0.140,"mRNA metabolism"),
c("GO:0006396","RNA processing",3.210,47.1864,0.819,0.373,"mRNA metabolism"),
c("GO:0008380","RNA splicing",0.413,28.1379,0.823,0.299,"mRNA metabolism"),
c("GO:0031647","regulation of protein stability",0.070,3.2076,0.868,0.411,"mRNA metabolism"),
c("GO:1901361","organic cyclic compound catabolic process",1.164,15.6003,0.818,0.665,"mRNA metabolism"),
c("GO:0034248","regulation of cellular amide metabolic process",0.700,11.6925,0.752,0.488,"mRNA metabolism"),
c("GO:0033238","regulation of cellular amine metabolic process",0.021,10.7670,0.809,0.631,"mRNA metabolism"),
c("GO:0031123","RNA 3'-end processing",0.145,4.0788,0.836,0.554,"mRNA metabolism"),
c("GO:0098781","ncRNA transcription",0.045,7.4535,0.856,0.528,"mRNA metabolism"),
c("GO:0009452","7-methylguanosine RNA capping",0.112,4.0137,0.837,0.542,"mRNA metabolism"),
c("GO:0042795","snRNA transcription from RNA polymerase II promoter",0.003,5.1637,0.867,0.481,"mRNA metabolism"),
c("GO:0034660","ncRNA metabolic process",3.407,35.8125,0.826,0.453,"mRNA metabolism"),
c("GO:0022613","ribonucleoprotein complex biogenesis",1.614,52.3316,0.807,0.000,"ribonucleoprotein complex biogenesis"),
c("GO:0071824","protein-DNA complex subunit organization",0.238,3.8508,0.799,0.629,"ribonucleoprotein complex biogenesis"),
c("GO:0071826","ribonucleoprotein complex subunit organization",0.377,24.2984,0.792,0.463,"ribonucleoprotein complex biogenesis"),
c("GO:0016050","vesicle organization",0.130,9.5100,0.812,0.474,"ribonucleoprotein complex biogenesis"),
c("GO:0043248","proteasome assembly",0.049,4.3768,0.796,0.664,"ribonucleoprotein complex biogenesis"),
c("GO:0061024","membrane organization",0.759,24.4225,0.806,0.499,"ribonucleoprotein complex biogenesis"),
c("GO:0051276","chromosome organization",1.477,6.5686,0.775,0.684,"ribonucleoprotein complex biogenesis"),
c("GO:0006325","chromatin organization",0.668,3.5986,0.783,0.690,"ribonucleoprotein complex biogenesis"),
c("GO:0006900","membrane budding",0.030,16.1421,0.737,0.367,"ribonucleoprotein complex biogenesis"),
c("GO:0070271","protein complex biogenesis",0.963,7.3179,0.815,0.684,"ribonucleoprotein complex biogenesis"),
c("GO:0042147","retrograde transport, endosome to Golgi",0.055,4.6253,0.823,0.671,"ribonucleoprotein complex biogenesis"),
c("GO:0031468","nuclear envelope reassembly",0.003,3.3420,0.810,0.689,"ribonucleoprotein complex biogenesis"),
c("GO:1904896","ESCRT complex disassembly",0.000,3.9101,0.852,0.401,"ribonucleoprotein complex biogenesis"),
c("GO:0044087","regulation of cellular component biogenesis",0.404,6.9208,0.717,0.622,"ribonucleoprotein complex biogenesis"),
c("GO:1904903","ESCRT III complex disassembly",0.000,3.9101,0.852,0.247,"ribonucleoprotein complex biogenesis"),
c("GO:1901673","regulation of mitotic spindle assembly",0.005,3.3449,0.685,0.684,"ribonucleoprotein complex biogenesis"),
c("GO:0000028","ribosomal small subunit assembly",0.020,5.2125,0.791,0.518,"ribonucleoprotein complex biogenesis"),
c("GO:0070925","organelle assembly",0.571,8.8928,0.766,0.645,"ribonucleoprotein complex biogenesis"),
c("GO:0016570","histone modification",0.373,3.5017,0.679,0.654,"ribonucleoprotein complex biogenesis"),
c("GO:0007030","Golgi organization",0.039,3.6576,0.819,0.435,"ribonucleoprotein complex biogenesis"),
c("GO:0007033","vacuole organization",0.102,11.5702,0.815,0.428,"ribonucleoprotein complex biogenesis"),
c("GO:1990182","exosomal secretion",0.004,4.8697,0.766,0.493,"ribonucleoprotein complex biogenesis"),
c("GO:0010256","endomembrane system organization",0.189,7.3019,0.825,0.432,"ribonucleoprotein complex biogenesis"),
c("GO:0048193","Golgi vesicle transport",0.297,14.3635,0.883,0.643,"ribonucleoprotein complex biogenesis"),
c("GO:0043254","regulation of protein complex assembly",0.198,8.4001,0.668,0.619,"ribonucleoprotein complex biogenesis"),
c("GO:0051351","positive regulation of ligase activity",0.002,15.7033,0.863,0.000,"positive regulation of ligase activity"),
c("GO:0032781","positive regulation of ATPase activity",0.033,3.6556,0.844,0.543,"positive regulation of ligase activity"),
c("GO:0051340","regulation of ligase activity",0.002,14.3298,0.867,0.375,"positive regulation of ligase activity"),
c("GO:0051338","regulation of transferase activity",0.368,9.9586,0.821,0.646,"positive regulation of ligase activity"),
c("GO:0051348","negative regulation of transferase activity",0.088,13.5003,0.822,0.456,"positive regulation of ligase activity"),
c("GO:0044033","multi-organism metabolic process",0.025,35.6655,0.928,0.011,"multi-organism metabolism"),
c("GO:1902590","multi-organism organelle organization",0.006,5.4868,0.795,0.608,"multi-organism metabolism"),
c("GO:0044764","multi-organism cellular process",0.325,26.2132,0.892,0.692,"multi-organism metabolism"),
c("GO:0044803","multi-organism membrane organization",0.008,4.4989,0.790,0.621,"multi-organism metabolism"),
c("GO:0019083","viral transcription",0.012,33.5114,0.809,0.562,"multi-organism metabolism"),
c("GO:0042026","protein refolding",0.069,7.2269,0.959,0.027,"protein refolding"),
c("GO:0061077","chaperone-mediated protein folding",0.043,4.8697,0.960,0.621,"protein refolding"),
c("GO:0016236","macroautophagy",0.084,10.4895,0.944,0.027,"macroautophagy"),
c("GO:0006914","autophagy",0.230,13.1314,0.959,0.030,"autophagy"),
c("GO:0000278","mitotic cell cycle",0.561,17.3307,0.873,0.033,"mitotic cell cycle"),
c("GO:0034976","response to endoplasmic reticulum stress",0.100,9.0353,0.885,0.263,"mitotic cell cycle"),
c("GO:0035966","response to topologically incorrect protein",0.053,8.2248,0.914,0.407,"mitotic cell cycle"),
c("GO:0051301","cell division",1.230,4.1018,0.912,0.223,"mitotic cell cycle"),
c("GO:0060071","Wnt signaling pathway, planar cell polarity pathway",0.010,10.7620,0.771,0.130,"mitotic cell cycle"),
c("GO:0000920","cell separation after cytokinesis",0.013,4.2741,0.935,0.132,"mitotic cell cycle"),
c("GO:0007164","establishment of tissue polarity",0.019,9.2581,0.965,0.462,"mitotic cell cycle"),
c("GO:0070911","global genome nucleotide-excision repair",0.003,3.0980,0.844,0.624,"mitotic cell cycle"),
c("GO:0007049","cell cycle",1.885,15.0031,0.909,0.204,"mitotic cell cycle"),
c("GO:0007017","microtubule-based process",0.658,5.7852,0.916,0.208,"mitotic cell cycle"),
c("GO:0006986","response to unfolded protein",0.037,8.3116,0.908,0.605,"mitotic cell cycle"),
c("GO:0038061","NIK/NF-kappaB signaling",0.018,8.6556,0.821,0.284,"mitotic cell cycle"),
c("GO:0033209","tumor necrosis factor-mediated signaling pathway",0.016,5.6364,0.784,0.573,"mitotic cell cycle"),
c("GO:0000715","nucleotide-excision repair, DNA damage recognition",0.033,8.3458,0.701,0.457,"mitotic cell cycle"),
c("GO:0071495","cellular response to endogenous stimulus",0.402,4.4248,0.932,0.297,"mitotic cell cycle"),
c("GO:0006457","protein folding",0.903,14.9031,0.955,0.034,"protein folding"),
c("GO:0016311","dephosphorylation",1.250,3.7986,0.931,0.065,"dephosphorylation"));


#### Cell > 4*Nuc (exons only) ####
revigo.data.list[["CC_cell_exons"]] <- rbind(c("GO:0005730","nucleolus",0.664,36.1586,0.516,0.000,"nucleolus"),
c("GO:0005694","chromosome",1.505,5.5258,0.643,0.633,"nucleolus"),
c("GO:0005697","telomerase holoenzyme complex",0.025,1.7258,0.543,0.505,"nucleolus"),
c("GO:0090575","RNA polymerase II transcription factor complex",0.155,4.2510,0.482,0.585,"nucleolus"),
c("GO:0008023","transcription elongation factor complex",0.080,1.5171,0.487,0.662,"nucleolus"),
c("GO:0033290","eukaryotic 48S preinitiation complex",0.089,7.8297,0.601,0.419,"nucleolus"),
c("GO:0089701","U2AF",0.017,1.3063,0.545,0.491,"nucleolus"),
c("GO:0000346","transcription export complex",0.014,1.6819,0.550,0.483,"nucleolus"),
c("GO:0022626","cytosolic ribosome",0.187,30.0438,0.530,0.413,"nucleolus"),
c("GO:0005635","nuclear envelope",0.283,3.5072,0.504,0.617,"nucleolus"),
c("GO:0005681","spliceosomal complex",0.250,21.0164,0.479,0.610,"nucleolus"),
c("GO:0071204","histone pre-mRNA 3'end processing complex",0.001,2.2097,0.604,0.416,"nucleolus"),
c("GO:0000803","sex chromosome",0.007,1.8182,0.727,0.408,"nucleolus"),
c("GO:0005662","DNA replication factor A complex",0.009,2.3063,0.527,0.515,"nucleolus"),
c("GO:0097346","INO80-type complex",0.056,2.4935,0.511,0.538,"nucleolus"),
c("GO:0036019","endolysosome",0.001,1.5243,0.615,0.544,"nucleolus"),
c("GO:0036020","endolysosome membrane",0.000,1.9066,0.580,0.523,"nucleolus"),
c("GO:0000805","X chromosome",0.002,2.3224,0.742,0.295,"nucleolus"),
c("GO:0031428","box C/D snoRNP complex",0.012,1.3063,0.543,0.643,"nucleolus"),
c("GO:0005840","ribosome",4.198,15.4001,0.433,0.589,"nucleolus"),
c("GO:0000407","pre-autophagosomal structure",0.033,2.9431,0.751,0.247,"nucleolus"),
c("GO:0036464","cytoplasmic ribonucleoprotein granule",0.127,6.7878,0.548,0.430,"nucleolus"),
c("GO:0097422","tubular endosome",0.002,1.6421,0.582,0.574,"nucleolus"),
c("GO:0008180","COP9 signalosome",0.031,2.0101,0.530,0.513,"nucleolus"),
c("GO:0099568","cytoplasmic region",0.265,5.1599,0.716,0.288,"nucleolus"),
c("GO:0015934","large ribosomal subunit",0.559,12.4023,0.475,0.666,"nucleolus"),
c("GO:0005768","endosome",0.319,14.3391,0.508,0.617,"nucleolus"),
c("GO:0031965","nuclear membrane",0.100,1.3497,0.516,0.676,"nucleolus"),
c("GO:0005776","autophagosome",0.024,2.4283,0.668,0.256,"nucleolus"),
c("GO:0005773","vacuole",0.455,11.7620,0.629,0.329,"nucleolus"),
c("GO:0070822","Sin3-type complex",0.010,1.5243,0.506,0.581,"nucleolus"),
c("GO:0031984","organelle subcompartment",0.269,5.7375,0.620,0.402,"nucleolus"),
c("GO:0005801","cis-Golgi network",0.038,2.7055,0.721,0.266,"nucleolus"),
c("GO:0048500","signal recognition particle",0.100,2.2097,0.602,0.422,"nucleolus"),
c("GO:0005785","signal recognition particle receptor complex",0.011,2.1013,0.511,0.598,"nucleolus"),
c("GO:0048471","perinuclear region of cytoplasm",0.135,8.9281,0.728,0.114,"nucleolus"),
c("GO:0005793","endoplasmic reticulum-Golgi intermediate compartment",0.026,6.7235,0.690,0.258,"nucleolus"),
c("GO:0005794","Golgi apparatus",0.969,13.5243,0.532,0.684,"nucleolus"),
c("GO:0048770","pigment granule",0.016,3.4841,0.615,0.674,"nucleolus"),
c("GO:0070993","translation preinitiation complex",0.093,7.0985,0.603,0.420,"nucleolus"),
c("GO:0097525","spliceosomal snRNP complex",0.128,8.2565,0.486,0.575,"nucleolus"),
c("GO:0044452","nucleolar part",0.153,7.8827,0.556,0.642,"nucleolus"),
c("GO:0044445","cytosolic part",0.496,24.4283,0.695,0.617,"nucleolus"),
c("GO:0031519","PcG protein complex",0.017,1.9914,0.545,0.491,"nucleolus"),
c("GO:0035770","ribonucleoprotein granule",0.131,6.3215,0.583,0.431,"nucleolus"),
c("GO:0044427","chromosomal part",1.117,4.8928,0.548,0.608,"nucleolus"),
c("GO:0071013","catalytic step 2 spliceosome",0.062,17.1129,0.477,0.541,"nucleolus"),
c("GO:0030532","small nuclear ribonucleoprotein complex",0.138,8.5918,0.497,0.579,"nucleolus"),
c("GO:0016592","mediator complex",0.155,5.2899,0.468,0.694,"nucleolus"),
c("GO:0033116","endoplasmic reticulum-Golgi intermediate compartment membrane",0.008,5.1568,0.624,0.587,"nucleolus"),
c("GO:0032588","trans-Golgi network membrane",0.008,2.4283,0.577,0.679,"nucleolus"),
c("GO:0034045","pre-autophagosomal structure membrane",0.018,1.6946,0.708,0.237,"nucleolus"),
c("GO:0016607","nuclear speck",0.091,7.3788,0.571,0.614,"nucleolus"),
c("GO:0015630","microtubule cytoskeleton",0.900,11.7620,0.631,0.592,"nucleolus"),
c("GO:0000139","Golgi membrane",0.403,17.6576,0.475,0.418,"nucleolus"),
c("GO:0030684","preribosome",0.223,6.7235,0.644,0.600,"nucleolus"),
c("GO:0070603","SWI/SNF superfamily-type complex",0.112,3.2000,0.495,0.569,"nucleolus"),
c("GO:0005881","cytoplasmic microtubule",0.023,2.0799,0.604,0.628,"nucleolus"),
c("GO:0005885","Arp2/3 protein complex",0.063,4.0640,0.529,0.680,"nucleolus"),
c("GO:0042470","melanosome",0.016,3.4841,0.615,0.673,"nucleolus"),
c("GO:0035145","exon-exon junction complex",0.010,1.6326,0.557,0.473,"nucleolus"),
c("GO:0005925","focal adhesion",0.109,9.5031,0.972,0.000,"focal adhesion"),
c("GO:0030054","cell junction",0.445,3.5670,0.994,0.000,"cell junction"),
c("GO:0044456","synapse part",0.230,8.4306,0.949,0.000,"synapse part"),
c("GO:0045202","synapse",0.299,7.4522,0.994,0.000,"synapse"),
c("GO:0048475","coated membrane",0.264,9.5031,0.984,0.000,"coated membrane"),
c("GO:1902494","catalytic complex",3.734,36.0292,0.689,0.000,"catalytic complex"),
c("GO:0070765","gamma-secretase complex",0.003,1.4841,0.718,0.438,"catalytic complex"),
c("GO:0016272","prefoldin complex",0.056,2.7747,0.721,0.383,"catalytic complex"),
c("GO:0000307","cyclin-dependent protein kinase holoenzyme complex",0.027,2.2757,0.595,0.610,"catalytic complex"),
c("GO:0000306","extrinsic component of vacuolar membrane",0.005,2.3224,0.601,0.677,"catalytic complex"),
c("GO:0016281","eukaryotic translation initiation factor 4F complex",0.002,1.9872,0.668,0.243,"catalytic complex"),
c("GO:0005732","small nucleolar ribonucleoprotein complex",0.072,4.4283,0.668,0.321,"catalytic complex"),
c("GO:0032299","ribonuclease H2 complex",0.009,1.4841,0.697,0.279,"catalytic complex"),
c("GO:0036513","Derlin-1 retrotranslocation complex",0.005,3.0022,0.548,0.457,"catalytic complex"),
c("GO:0031248","protein acetyltransferase complex",0.152,3.2299,0.591,0.693,"catalytic complex"),
c("GO:1990124","messenger ribonucleoprotein complex",0.002,1.6968,0.725,0.243,"catalytic complex"),
c("GO:0090498","extrinsic component of Golgi membrane",0.001,1.4841,0.607,0.693,"catalytic complex"),
c("GO:0005834","heterotrimeric G-protein complex",0.046,2.3778,0.585,0.530,"catalytic complex"),
c("GO:0005844","polysome",0.037,2.3325,0.681,0.368,"catalytic complex"),
c("GO:0000974","Prp19 complex",0.017,4.7305,0.740,0.289,"catalytic complex"),
c("GO:0032937","SREBP-SCAP-Insig complex",0.001,1.4841,0.583,0.411,"catalytic complex"),
c("GO:0030904","retromer complex",0.032,4.1180,0.596,0.516,"catalytic complex"),
c("GO:1903293","phosphatase complex",0.073,9.3188,0.700,0.545,"catalytic complex"),
c("GO:0030906","retromer, cargo-selective complex",0.002,2.1013,0.645,0.434,"catalytic complex"),
c("GO:0005851","eukaryotic translation initiation factor 2B complex",0.009,1.3063,0.646,0.279,"catalytic complex"),
c("GO:0005852","eukaryotic translation initiation factor 3 complex",0.117,7.0985,0.590,0.405,"catalytic complex"),
c("GO:0005853","eukaryotic translation elongation factor 1 complex",0.012,2.1013,0.640,0.290,"catalytic complex"),
c("GO:0031461","cullin-RING ubiquitin ligase complex",0.159,9.0022,0.587,0.695,"catalytic complex"),
c("GO:0030877","beta-catenin destruction complex",0.004,1.5003,0.658,0.260,"catalytic complex"),
c("GO:0005786","signal recognition particle, endoplasmic reticulum targeting",0.034,2.2097,0.626,0.366,"catalytic complex"),
c("GO:0034708","methyltransferase complex",0.080,7.8297,0.612,0.659,"catalytic complex"),
c("GO:0072669","tRNA-splicing ligase complex",0.006,2.3224,0.703,0.273,"catalytic complex"),
c("GO:1990904","ribonucleoprotein complex",5.291,56.4260,0.680,0.543,"catalytic complex"),
c("GO:0070971","endoplasmic reticulum exit site",0.006,2.1746,0.614,0.486,"catalytic complex"),
c("GO:0031501","mannosyltransferase complex",0.014,1.5100,0.648,0.478,"catalytic complex"),
c("GO:0071541","eukaryotic translation initiation factor 3 complex, eIF3m",0.011,4.7423,0.641,0.273,"catalytic complex"),
c("GO:0002199","zona pellucida receptor complex",0.003,2.1746,0.764,0.244,"catalytic complex"),
c("GO:0030117","membrane coat",0.264,9.5031,0.534,0.366,"catalytic complex"),
c("GO:0034709","methylosome",0.006,7.4101,0.607,0.447,"catalytic complex"),
c("GO:1902911","protein kinase complex",0.089,1.7011,0.660,0.671,"catalytic complex"),
c("GO:0034715","pICln-Sm protein complex",0.006,3.3893,0.653,0.260,"catalytic complex"),
c("GO:0000502","proteasome complex",0.389,16.8894,0.551,0.636,"catalytic complex"),
c("GO:0034719","SMN-Sm protein complex",0.008,2.7399,0.647,0.278,"catalytic complex"),
c("GO:0008287","protein serine/threonine phosphatase complex",0.073,9.3188,0.665,0.545,"catalytic complex"),
c("GO:0034098","VCP-NPL4-UFD1 AAA ATPase complex",0.005,2.8665,0.548,0.491,"catalytic complex"),
c("GO:0033176","proton-transporting V-type ATPase complex",0.146,1.6326,0.681,0.579,"catalytic complex"),
c("GO:0033185","dolichol-phosphate-mannose synthase complex",0.002,1.4841,0.685,0.236,"catalytic complex"),
c("GO:0030687","preribosome, large subunit precursor",0.067,4.3233,0.663,0.383,"catalytic complex"),
c("GO:0030677","ribonuclease P complex",0.021,1.5100,0.631,0.492,"catalytic complex"),
c("GO:0033202","DNA helicase complex",0.093,1.8962,0.637,0.556,"catalytic complex"),
c("GO:0031201","SNARE complex",0.038,5.0635,0.600,0.523,"catalytic complex"),
c("GO:0000151","ubiquitin ligase complex",0.232,13.8665,0.586,0.604,"catalytic complex"),
c("GO:0098796","membrane protein complex",2.473,6.7932,0.635,0.533,"catalytic complex"),
c("GO:0097255","R2TP complex",0.003,2.7100,0.763,0.247,"catalytic complex"),
c("GO:1990234","transferase complex",1.223,33.5058,0.638,0.439,"catalytic complex"),
c("GO:0042788","polysomal ribosome",0.010,1.5100,0.611,0.339,"catalytic complex"),
c("GO:1902493","acetyltransferase complex",0.152,3.2299,0.655,0.693,"catalytic complex"),
c("GO:0019898","extrinsic component of membrane",0.249,2.7878,0.982,0.031,"extrinsic component of membrane"),
c("GO:0030496","midbody",0.040,4.4437,0.922,0.048,"midbody"),
c("GO:0043209","myelin sheath",0.049,8.7932,0.921,0.049,"myelin sheath"),
c("GO:0031252","cell leading edge",0.086,4.4828,0.918,0.051,"cell leading edge"),
c("GO:0044297","cell body",0.087,7.0964,0.918,0.051,"cell body"),
c("GO:0030427","site of polarized growth",0.091,7.6635,0.918,0.051,"site of polarized growth"),
c("GO:0043005","neuron projection",0.190,21.0575,0.820,0.055,"neuron projection"),
c("GO:0032838","cell projection cytoplasm",0.014,4.6696,0.705,0.595,"neuron projection"),
c("GO:0035253","ciliary rootlet",0.003,1.6968,0.640,0.543,"neuron projection"),
c("GO:0030027","lamellipodium",0.040,2.6144,0.867,0.640,"neuron projection"),
c("GO:0097458","neuron part",0.320,18.7959,0.911,0.057,"neuron part"),
c("GO:0016234","inclusion body",0.019,1.8182,0.857,0.098,"inclusion body"));

  
revigo.data.list[["BP_cell_exons"]] <- rbind(c("GO:0015031","protein transport",2.251,45.5258,0.845,0.000,"protein transport"),
c("GO:0071166","ribonucleoprotein complex localization",0.097,5.1805,0.841,0.587,"protein transport"),
c("GO:0006900","membrane budding",0.030,17.7100,0.738,0.643,"protein transport"),
c("GO:1990778","protein localization to cell periphery",0.063,4.3665,0.834,0.635,"protein transport"),
c("GO:0042147","retrograde transport, endosome to Golgi",0.055,4.3575,0.820,0.671,"protein transport"),
c("GO:0016192","vesicle-mediated transport",1.085,17.7799,0.909,0.344,"protein transport"),
c("GO:0016197","endosomal transport",0.131,13.7825,0.922,0.274,"protein transport"),
c("GO:1903827","regulation of cellular protein localization",0.132,7.6326,0.700,0.680,"protein transport"),
c("GO:0050657","nucleic acid transport",0.100,3.4776,0.906,0.533,"protein transport"),
c("GO:1904874","positive regulation of telomerase RNA localization to Cajal body",0.004,3.6990,0.741,0.484,"protein transport"),
c("GO:0048278","vesicle docking",0.080,3.4647,0.835,0.691,"protein transport"),
c("GO:0007034","vacuolar transport",0.133,14.1002,0.922,0.275,"protein transport"),
c("GO:0007041","lysosomal transport",0.017,3.8013,0.928,0.229,"protein transport"),
c("GO:0030705","cytoskeleton-dependent intracellular transport",0.056,5.1561,0.841,0.605,"protein transport"),
c("GO:0006403","RNA localization",0.118,6.8297,0.882,0.643,"protein transport"),
c("GO:0048193","Golgi vesicle transport",0.297,19.6234,0.887,0.298,"protein transport"),
c("GO:0051668","localization within membrane",0.023,4.7235,0.855,0.529,"protein transport"),
c("GO:0006913","nucleocytoplasmic transport",0.237,12.1273,0.817,0.291,"protein transport"),
c("GO:0051640","organelle localization",0.223,9.0083,0.831,0.627,"protein transport"),
c("GO:0051650","establishment of vesicle localization",0.072,9.4134,0.826,0.605,"protein transport"),
c("GO:0016482","cytosolic transport",0.076,8.6615,0.837,0.619,"protein transport"),
c("GO:0051169","nuclear transport",0.239,12.4486,0.824,0.677,"protein transport"),
c("GO:0016071","mRNA metabolic process",0.798,52.6326,0.851,0.000,"mRNA metabolism"),
c("GO:0045862","positive regulation of proteolysis",0.078,6.6345,0.654,0.609,"mRNA metabolism"),
c("GO:0031329","regulation of cellular catabolic process",0.093,17.0814,0.701,0.583,"mRNA metabolism"),
c("GO:0006369","termination of RNA polymerase II transcription",0.015,4.6925,0.871,0.526,"mRNA metabolism"),
c("GO:0006370","7-methylguanosine mRNA capping",0.022,6.0438,0.846,0.479,"mRNA metabolism"),
c("GO:0006367","transcription initiation from RNA polymerase II promoter",0.109,6.2596,0.854,0.601,"mRNA metabolism"),
c("GO:0006368","transcription elongation from RNA polymerase II promoter",0.082,11.7645,0.854,0.300,"mRNA metabolism"),
c("GO:0016073","snRNA metabolic process",0.035,5.4828,0.871,0.416,"mRNA metabolism"),
c("GO:0070647","protein modification by small protein conjugation or removal",0.821,30.0799,0.816,0.331,"mRNA metabolism"),
c("GO:0006354","DNA-templated transcription, elongation",0.202,11.5331,0.855,0.280,"mRNA metabolism"),
c("GO:0006353","DNA-templated transcription, termination",0.185,8.2388,0.856,0.320,"mRNA metabolism"),
c("GO:0006352","DNA-templated transcription, initiation",0.766,7.1701,0.839,0.361,"mRNA metabolism"),
c("GO:0044265","cellular macromolecule catabolic process",1.268,48.6421,0.761,0.153,"mRNA metabolism"),
c("GO:0000338","protein deneddylation",0.017,4.1057,0.815,0.683,"mRNA metabolism"),
c("GO:0043618","regulation of transcription from RNA polymerase II promoter in response to stress",0.026,3.8570,0.721,0.545,"mRNA metabolism"),
c("GO:0010608","posttranscriptional regulation of gene expression",0.719,23.5421,0.755,0.256,"mRNA metabolism"),
c("GO:1901566","organonitrogen compound biosynthetic process",14.064,8.4089,0.907,0.391,"mRNA metabolism"),
c("GO:0043603","cellular amide metabolic process",6.879,8.9318,0.892,0.189,"mRNA metabolism"),
c("GO:0010563","negative regulation of phosphorus metabolic process",0.182,4.4634,0.755,0.687,"mRNA metabolism"),
c("GO:0043543","protein acylation",0.202,3.8539,0.834,0.429,"mRNA metabolism"),
c("GO:0006259","DNA metabolic process",5.607,3.5986,0.836,0.383,"mRNA metabolism"),
c("GO:0051437","positive regulation of ubiquitin-protein ligase activity involved in regulation of mitotic cell cycle transition",0.001,18.0287,0.587,0.689,"mRNA metabolism"),
c("GO:0031400","negative regulation of protein modification process",0.151,15.5376,0.660,0.418,"mRNA metabolism"),
c("GO:0006521","regulation of cellular amino acid metabolic process",0.018,11.9747,0.787,0.276,"mRNA metabolism"),
c("GO:0043488","regulation of mRNA stability",0.018,15.0526,0.796,0.293,"mRNA metabolism"),
c("GO:0002181","cytoplasmic translation",0.064,9.2125,0.831,0.429,"mRNA metabolism"),
c("GO:0043487","regulation of RNA stability",0.021,14.2381,0.794,0.573,"mRNA metabolism"),
c("GO:0032535","regulation of cellular component size",0.179,4.5086,0.704,0.437,"mRNA metabolism"),
c("GO:0036260","RNA capping",0.112,5.3516,0.847,0.542,"mRNA metabolism"),
c("GO:0016567","protein ubiquitination",0.523,28.9957,0.777,0.470,"mRNA metabolism"),
c("GO:0006412","translation",5.686,18.7878,0.756,0.661,"mRNA metabolism"),
c("GO:0009894","regulation of catabolic process",0.146,18.7670,0.755,0.530,"mRNA metabolism"),
c("GO:0006413","translational initiation",0.518,37.7447,0.803,0.140,"mRNA metabolism"),
c("GO:0006396","RNA processing",3.210,47.2069,0.824,0.373,"mRNA metabolism"),
c("GO:0008380","RNA splicing",0.413,31.7167,0.832,0.299,"mRNA metabolism"),
c("GO:1901361","organic cyclic compound catabolic process",1.164,11.7773,0.825,0.665,"mRNA metabolism"),
c("GO:0034248","regulation of cellular amide metabolic process",0.700,12.3595,0.745,0.488,"mRNA metabolism"),
c("GO:0008334","histone mRNA metabolic process",0.005,3.6517,0.878,0.522,"mRNA metabolism"),
c("GO:0006470","protein dephosphorylation",0.585,4.9872,0.813,0.475,"mRNA metabolism"),
c("GO:0033238","regulation of cellular amine metabolic process",0.021,9.7905,0.803,0.631,"mRNA metabolism"),
c("GO:0031123","RNA 3'-end processing",0.145,3.5544,0.844,0.554,"mRNA metabolism"),
c("GO:0098781","ncRNA transcription",0.045,6.9914,0.860,0.287,"mRNA metabolism"),
c("GO:0000184","nuclear-transcribed mRNA catabolic process, nonsense-mediated decay",0.030,31.2807,0.770,0.619,"mRNA metabolism"),
c("GO:0031145","anaphase-promoting complex-dependent catabolic process",0.021,16.7773,0.774,0.600,"mRNA metabolism"),
c("GO:0042795","snRNA transcription from RNA polymerase II promoter",0.003,4.4535,0.871,0.481,"mRNA metabolism"),
c("GO:0034660","ncRNA metabolic process",3.407,28.5317,0.831,0.453,"mRNA metabolism"),
c("GO:0022613","ribonucleoprotein complex biogenesis",1.614,42.8297,0.809,0.000,"ribonucleoprotein complex biogenesis"),
c("GO:0016050","vesicle organization",0.130,13.2299,0.812,0.417,"ribonucleoprotein complex biogenesis"),
c("GO:0031346","positive regulation of cell projection organization",0.056,6.7670,0.655,0.637,"ribonucleoprotein complex biogenesis"),
c("GO:0043248","proteasome assembly",0.049,4.1561,0.801,0.510,"ribonucleoprotein complex biogenesis"),
c("GO:0006364","rRNA processing",0.952,32.3706,0.661,0.683,"ribonucleoprotein complex biogenesis"),
c("GO:0061024","membrane organization",0.759,24.2328,0.805,0.499,"ribonucleoprotein complex biogenesis"),
c("GO:0051276","chromosome organization",1.477,5.6198,0.774,0.673,"ribonucleoprotein complex biogenesis"),
c("GO:0070271","protein complex biogenesis",0.963,8.1141,0.817,0.684,"ribonucleoprotein complex biogenesis"),
c("GO:0008154","actin polymerization or depolymerization",0.153,4.3188,0.772,0.544,"ribonucleoprotein complex biogenesis"),
c("GO:0031468","nuclear envelope reassembly",0.003,3.8327,0.807,0.689,"ribonucleoprotein complex biogenesis"),
c("GO:0044087","regulation of cellular component biogenesis",0.404,7.9101,0.713,0.622,"ribonucleoprotein complex biogenesis"),
c("GO:0033043","regulation of organelle organization",0.495,8.2041,0.671,0.536,"ribonucleoprotein complex biogenesis"),
c("GO:0000028","ribosomal small subunit assembly",0.020,3.6478,0.800,0.540,"ribonucleoprotein complex biogenesis"),
c("GO:0007030","Golgi organization",0.039,5.7825,0.817,0.442,"ribonucleoprotein complex biogenesis"),
c("GO:0007033","vacuole organization",0.102,11.6635,0.815,0.474,"ribonucleoprotein complex biogenesis"),
c("GO:0030030","cell projection organization",0.608,4.9872,0.778,0.487,"ribonucleoprotein complex biogenesis"),
c("GO:0010256","endomembrane system organization",0.189,12.6904,0.824,0.432,"ribonucleoprotein complex biogenesis"),
c("GO:0044033","multi-organism metabolic process",0.025,27.3197,0.940,0.011,"multi-organism metabolism"),
c("GO:1902590","multi-organism organelle organization",0.006,4.4976,0.806,0.608,"multi-organism metabolism"),
c("GO:0044764","multi-organism cellular process",0.325,20.8097,0.905,0.692,"multi-organism metabolism"),
c("GO:0019083","viral transcription",0.012,25.1898,0.827,0.562,"multi-organism metabolism"),
c("GO:0038093","Fc receptor signaling pathway",0.005,16.2418,0.755,0.022,"Fc receptor signaling pathway"),
c("GO:0002478","antigen processing and presentation of exogenous peptide antigen",0.004,11.5986,0.922,0.505,"Fc receptor signaling pathway"),
c("GO:0034976","response to endoplasmic reticulum stress",0.100,11.2573,0.880,0.252,"Fc receptor signaling pathway"),
c("GO:0006283","transcription-coupled nucleotide-excision repair",0.038,9.9626,0.819,0.462,"Fc receptor signaling pathway"),
c("GO:0035966","response to topologically incorrect protein",0.053,9.4318,0.908,0.407,"Fc receptor signaling pathway"),
c("GO:0042059","negative regulation of epidermal growth factor receptor signaling pathway",0.007,6.8928,0.762,0.559,"Fc receptor signaling pathway"),
c("GO:0007264","small GTPase mediated signal transduction",0.485,4.9626,0.767,0.443,"Fc receptor signaling pathway"),
c("GO:0051340","regulation of ligase activity",0.002,14.4660,0.858,0.375,"Fc receptor signaling pathway"),
c("GO:0051338","regulation of transferase activity",0.368,11.8827,0.807,0.646,"Fc receptor signaling pathway"),
c("GO:0051348","negative regulation of transferase activity",0.088,14.1101,0.809,0.456,"Fc receptor signaling pathway"),
c("GO:0051351","positive regulation of ligase activity",0.002,15.7959,0.854,0.121,"Fc receptor signaling pathway"),
c("GO:0019882","antigen processing and presentation",0.027,7.6478,0.934,0.561,"Fc receptor signaling pathway"),
c("GO:0060071","Wnt signaling pathway, planar cell polarity pathway",0.010,9.6655,0.747,0.486,"Fc receptor signaling pathway"),
c("GO:0050690","regulation of defense response to virus by virus",0.005,4.1118,0.782,0.696,"Fc receptor signaling pathway"),
c("GO:0030518","intracellular steroid hormone receptor signaling pathway",0.024,4.1864,0.791,0.588,"Fc receptor signaling pathway"),
c("GO:0032107","regulation of response to nutrient levels",0.011,6.8729,0.820,0.576,"Fc receptor signaling pathway"),
c("GO:0006986","response to unfolded protein",0.037,8.9318,0.904,0.605,"Fc receptor signaling pathway"),
c("GO:0051098","regulation of binding",0.073,3.9547,0.828,0.560,"Fc receptor signaling pathway"),
c("GO:0006974","cellular response to DNA damage stimulus",2.360,7.8386,0.849,0.679,"Fc receptor signaling pathway"),
c("GO:0038061","NIK/NF-kappaB signaling",0.018,8.0931,0.811,0.273,"Fc receptor signaling pathway"),
c("GO:0080135","regulation of cellular response to stress",0.182,9.7670,0.748,0.548,"Fc receptor signaling pathway"),
c("GO:0033209","tumor necrosis factor-mediated signaling pathway",0.016,4.1765,0.774,0.573,"Fc receptor signaling pathway"),
c("GO:0042026","protein refolding",0.069,5.1959,0.958,0.027,"protein refolding"),
c("GO:0061077","chaperone-mediated protein folding",0.043,4.2396,0.959,0.621,"protein refolding"),
c("GO:0016236","macroautophagy",0.084,9.0353,0.943,0.027,"macroautophagy"),
c("GO:0006914","autophagy",0.230,10.5735,0.958,0.030,"autophagy"),
c("GO:0006457","protein folding",0.903,11.3233,0.954,0.034,"protein folding"),
c("GO:0007049","cell cycle",1.885,15.4976,0.900,0.059,"cell cycle"),
c("GO:0007017","microtubule-based process",0.658,5.1759,0.908,0.208,"cell cycle"),
c("GO:0007018","microtubule-based movement",0.287,4.3665,0.907,0.190,"cell cycle"),
c("GO:1903047","mitotic cell cycle process",0.514,14.9208,0.863,0.202,"cell cycle"),
c("GO:0022406","membrane docking",0.099,4.4962,0.919,0.172,"cell cycle"),
c("GO:0016311","dephosphorylation",1.250,5.4498,0.921,0.065,"dephosphorylation"),
c("GO:0007164","establishment of tissue polarity",0.019,7.8827,0.950,0.065,"establishment of tissue polarity"));
  
  
  
#### Nuc > 1.5*Cell (introns + exons) ####
revigo.data.list[["CC_nuc_ie"]] <- rbind(c("GO:0005911","cell-cell junction",0.222,1.4179,0.957,0.000,"cell-cell junction"),
c("GO:0030054","cell junction",0.445,5.9747,0.957,0.000,"cell junction"),
c("GO:0045202","synapse",0.299,10.4597,0.957,0.000,"synapse"),
c("GO:0098589","membrane region",0.121,5.0731,0.912,0.000,"membrane region"),
c("GO:0002095","caveolar macromolecular signaling complex",0.000,1.4112,0.694,0.410,"membrane region"),
c("GO:0098794","postsynapse",0.133,10.4597,0.682,0.000,"postsynapse"),
c("GO:0060076","excitatory synapse",0.004,6.5406,0.769,0.696,"postsynapse"),
c("GO:1902495","transmembrane transporter complex",0.864,4.7011,0.671,0.041,"transmembrane transporter complex"),
c("GO:0008328","ionotropic glutamate receptor complex",0.012,1.5751,0.535,0.681,"transmembrane transporter complex"),
c("GO:1990351","transporter complex",0.885,4.6383,0.832,0.357,"transmembrane transporter complex"),
c("GO:1990454","L-type voltage-gated calcium channel complex",0.001,2.3298,0.558,0.586,"transmembrane transporter complex"),
c("GO:0031225","anchored component of membrane",0.078,2.3233,0.928,0.043,"anchored component of membrane"),
c("GO:0044297","cell body",0.087,3.3497,0.899,0.045,"cell body"),
c("GO:0097458","neuron part",0.320,9.2620,0.894,0.050,"neuron part"),
c("GO:0043005","neuron projection",0.190,9.4179,0.503,0.051,"neuron projection"),
c("GO:0044224","juxtaparanode region of axon",0.002,1.8962,0.550,0.674,"neuron projection"),
c("GO:0032839","dendrite cytoplasm",0.003,1.3516,0.554,0.680,"neuron projection"),
c("GO:0098590","plasma membrane region",0.239,5.3595,0.645,0.052,"plasma membrane region"),
c("GO:0005887","integral component of plasma membrane",1.211,5.0237,0.603,0.696,"plasma membrane region"),
c("GO:0044431","Golgi apparatus part",0.608,1.3615,0.871,0.057,"Golgi apparatus part"));
  

revigo.data.list[["BP_nuc_ie"]] <- rbind(c("GO:0006935","chemotaxis",0.475,1.5346,0.925,0.000,"chemotaxis"),
c("GO:0007268","chemical synaptic transmission",0.187,4.9031,0.700,0.000,"chemical synaptic transmission"),
c("GO:0009214","cyclic nucleotide catabolic process",0.009,1.8996,0.829,0.698,"chemical synaptic transmission"),
c("GO:0008037","cell recognition",0.067,2.3507,0.844,0.137,"chemical synaptic transmission"),
c("GO:0031344","regulation of cell projection organization",0.123,3.4023,0.609,0.349,"chemical synaptic transmission"),
c("GO:0050808","synapse organization",0.070,4.7878,0.761,0.138,"chemical synaptic transmission"),
c("GO:0015014","heparan sulfate proteoglycan biosynthetic process, polysaccharide chain biosynthetic process",0.001,2.0655,0.864,0.105,"chemical synaptic transmission"),
c("GO:0007267","cell-cell signaling",0.407,2.6289,0.756,0.426,"chemical synaptic transmission"),
c("GO:0044089","positive regulation of cellular component biogenesis",0.193,1.6861,0.652,0.333,"chemical synaptic transmission"),
c("GO:0007196","adenylate cyclase-inhibiting G-protein coupled glutamate receptor signaling pathway",0.002,1.6696,0.765,0.297,"chemical synaptic transmission"),
c("GO:0051130","positive regulation of cellular component organization",0.374,1.5317,0.617,0.672,"chemical synaptic transmission"),
c("GO:0030030","cell projection organization",0.608,2.6364,0.727,0.413,"chemical synaptic transmission"),
c("GO:0006928","movement of cell or subcellular component",0.973,1.6073,0.815,0.171,"chemical synaptic transmission"),
c("GO:0006198","cAMP catabolic process",0.007,2.0186,0.826,0.247,"chemical synaptic transmission"),
c("GO:0007610","behavior",0.170,2.2757,0.983,0.000,"behavior"),
c("GO:0007626","locomotory behavior",0.061,1.6676,0.983,0.000,"locomotory behavior"),
c("GO:0022610","biological adhesion",0.550,3.4023,0.983,0.000,"biological adhesion"),
c("GO:0035418","protein localization to synapse",0.006,1.6861,0.959,0.000,"protein localization to synapse"),
c("GO:0097120","receptor localization to synapse",0.004,1.3401,0.959,0.138,"protein localization to synapse"),
c("GO:0034765","regulation of ion transmembrane transport",0.197,1.6778,0.760,0.579,"protein localization to synapse"),
c("GO:1901841","regulation of high voltage-gated calcium channel activity",0.000,1.4597,0.822,0.664,"protein localization to synapse"),
c("GO:0034220","ion transmembrane transport",3.528,1.5186,0.930,0.210,"protein localization to synapse"),
c("GO:0040011","locomotion",0.997,1.4522,0.983,0.000,"locomotion"),
c("GO:0050803","regulation of synapse structure or activity",0.034,6.6925,0.796,0.000,"regulation of synapse structure or activity"),
c("GO:0050905","neuromuscular process",0.023,1.5969,0.575,0.609,"regulation of synapse structure or activity"),
c("GO:0008038","neuron recognition",0.012,3.4841,0.441,0.664,"regulation of synapse structure or activity"),
c("GO:0048589","developmental growth",0.177,1.5186,0.638,0.553,"regulation of synapse structure or activity"),
c("GO:0051960","regulation of nervous system development",0.172,6.6696,0.340,0.176,"regulation of synapse structure or activity"),
c("GO:0061588","calcium activated phospholipid scrambling",0.004,1.9136,0.691,0.359,"regulation of synapse structure or activity"),
c("GO:0019226","transmission of nerve impulse",0.013,2.1624,0.491,0.514,"regulation of synapse structure or activity"),
c("GO:0010842","retina layer formation",0.006,1.8539,0.535,0.535,"regulation of synapse structure or activity"),
c("GO:0001508","action potential",0.024,2.0177,0.783,0.400,"regulation of synapse structure or activity"),
c("GO:0042391","regulation of membrane potential",0.135,3.2132,0.781,0.448,"regulation of synapse structure or activity"),
c("GO:0035637","multicellular organismal signaling",0.025,3.4023,0.536,0.536,"regulation of synapse structure or activity"),
c("GO:0031646","positive regulation of neurological system process",0.003,1.7190,0.461,0.630,"regulation of synapse structure or activity"),
c("GO:0060322","head development",0.151,2.0680,0.637,0.557,"regulation of synapse structure or activity"),
c("GO:1901078","negative regulation of relaxation of muscle",0.001,2.6459,0.493,0.556,"regulation of synapse structure or activity"),
c("GO:0032990","cell part morphogenesis",0.174,1.6073,0.547,0.563,"regulation of synapse structure or activity"),
c("GO:0097264","self proteolysis",0.040,3.2291,0.980,0.000,"self proteolysis"),
c("GO:0098742","cell-cell adhesion via plasma-membrane adhesion molecules",0.108,8.3468,0.928,0.000,"cell-cell adhesion via plasma-membrane adhesion molecules"));


#### Nuc > 1.5*Cell (exons) ####
revigo.data.list[["CC_nuc_exons"]] <- rbind(c("GO:0016607","nuclear speck",0.091,6.8697,0.217,0.000,"nuclear speck"),
c("GO:0005681","spliceosomal complex",0.250,4.5031,0.153,0.517,"nuclear speck"),
c("GO:0097525","spliceosomal snRNP complex",0.128,1.4802,0.141,0.531,"nuclear speck"),
c("GO:0030532","small nuclear ribonucleoprotein complex",0.138,1.3737,0.171,0.534,"nuclear speck"),
c("GO:0071004","U2-type prespliceosome",0.029,2.7773,0.164,0.443,"nuclear speck"),
c("GO:0030529","intracellular ribonucleoprotein complex",5.291,1.6326,0.222,0.624,"nuclear speck"),
c("GO:1990904","ribonucleoprotein complex",5.291,1.6326,0.421,0.000,"ribonucleoprotein complex"));


revigo.data.list[["BP_nuc_exons"]] <- rbind(c("GO:0008380","RNA splicing",0.413,9.0996,0.149,0.000,"RNA splicing"),
c("GO:0006396","RNA processing",3.210,4.4989,0.326,0.373,"RNA splicing"),
c("GO:0016071","mRNA metabolic process",0.798,4.7190,0.364,0.299,"RNA splicing"),
c("GO:0043484","regulation of RNA splicing",0.040,2.9914,0.089,0.500,"RNA splicing"),
c("GO:0006397","mRNA processing",0.561,6.1403,0.088,0.624,"RNA splicing"));


# A treemap R script produced by the REVIGO server at http://revigo.irb.hr/
# If you found REVIGO useful in your work, please cite the following reference:
# Supek F et al. "REVIGO summarizes and visualizes long lists of Gene Ontology
# terms" PLoS ONE 2011. doi:10.1371/journal.pone.0021800

# author: Anton Kratz <anton.kratz@gmail.com>, RIKEN Omics Science Center, Functional Genomics Technology Team, Japan
# created: Fri, Nov 02, 2012  7:25:52 PM
# last change: Fri, Nov 09, 2012  3:20:01 PM



# Treemap plots
revigo.names <- c("term_ID","description","freqInDbPercent","abslog10pvalue","uniqueness","dispensability","representative");

pdf( file="../analysis/nuc_vs_cell/nuc_vs_cell_go_enrich_revigo_treemaps.pdf", width=7, height=4) # width and height are in inches

for (set1 in names(revigo.data.list)) {
  revigo.data <- revigo.data.list[[set1]]
  
  stuff <- data.frame(revigo.data);
  names(stuff) <- revigo.names;
  
  stuff$abslog10pvalue <- as.numeric( as.character(stuff$abslog10pvalue) );
  stuff$freqInDbPercent <- as.numeric( as.character(stuff$freqInDbPercent) );
  stuff$uniqueness <- as.numeric( as.character(stuff$uniqueness) );
  stuff$dispensability <- as.numeric( as.character(stuff$dispensability) );
  
  # check the tmPlot command documentation for all possible parameters - there are a lot more
  treemap(
    stuff,
    index = c("representative","description"),
    vSize = "abslog10pvalue",
    type = "categorical",
    vColor = "representative",
    title = set1,
    inflate.labels = FALSE,      # set this to TRUE for space-filling group labels - good for posters
    lowerbound.cex.labels = 0,   # try to draw as many labels as possible (still, some small squares may not get a label)
    bg.labels = "#CCCCCCAA",     # define background color of group labels
    # "#CCCCCC00" is fully transparent, "#CCCCCCAA" is semi-transparent grey, NA is opaque
    position.legend = "none"
  )
}
dev.off()

```





```{r extra-code}
gene.prop.dex2l <- melt(gene.prop.dex2[, c("gene", "logFC", "gene_len", "AveExpr")],
                        id = c("gene", "logFC"))

g1b <- ggplot(gene.prop.dex2l, aes(x = logFC, y = value)) +
  facet_wrap(~ variable, scales = "free", ncol = 1) +
  # geom_vline(xintercept = 0, size = 1, color = "grey") +
  geom_point(alpha= 0.1) +
  # geom_vline(xintercept = c(-log2(1.5), log2(1.5)), size = 1, color = "grey") +
  # geom_text(data=cell.long.genes, aes(label = gene), 
  #           color = "black", hjust = 0, nudge_x = 0.1) +
  xlab("Nuclei vs. Cells (log2 fold change)") +
  ylab("Genomic length (base pairs)") +
  scale_color_gradientn(colors=brewer.pal(9, "Reds")) +
  theme_bw()
# plot(g1b)

g2 <- ggplot(gene.prop.dex2, aes(x = logFC_bin, y = gene_len)) +
  geom_hline(yintercept = median(gene.prop.dex2$gene_len)) +
  geom_boxplot() +
  scale_y_log10() +
  scale_x_discrete(breaks=seq(-2, 5.5, 0.5)) +
  theme_bw()
plot(g2)

# ggsave(g2, width = 6, height = 3,
#          filename = "../analysis/nuc_vs_cell/gene_length_vs_nuc_enrichment_boxplots.pdf")


gene.prop.dex2l <- melt(gene.prop.dex2[gene.prop.dex2$transcript_len < 1e5, 
                                       c("gene", "logFC", "gene_len", 
                                         "transcript_len")],
                        id = c("gene", "logFC"))

g3 <- ggplot(gene.prop.dex2l, aes(x = logFC, y = value)) +
  facet_wrap(~ variable, ncol = 1, scales = "free") +
  geom_vline(xintercept = 0) +
  geom_point() +
  ylab("Length (base pairs)") +
  theme_bw()
# plot(g3)


gene.prop.dex2$AveExpr_bin <- cut(gene.prop.dex2$AveExpr, breaks = 0:15)
gene.prop.dex2$gene_len_bin <- cut(gene.prop.dex2$gene_len, breaks = 10^seq(0, 7, by = 0.1))

g2b <- ggplot(gene.prop.dex2, aes(x = AveExpr_bin, y = logFC)) +
  # geom_hline(yintercept = median(gene.prop.dex2$gene_len)) +
  geom_boxplot() +
  coord_flip() +
  # scale_y_log10() +
  # scale_x_discrete(breaks=seq(-2, 5.5, 0.5)) +
  theme_bw()
plot(g2b)

```

