---
title: "Cell type homology"
output: html_notebook
---


```{r init-workspace, echo=FALSE, message=FALSE, warning=FALSE}
# Load libraries 
library(feather)
library(reshape2)
library(ggplot2)
library(ggrepel)
library(RColorBrewer)
library(pheatmap)
library(boot)
library(limma)

# Load functions
calc_beta <- function(y, spec.exp = 2) {
  d1 <- as.matrix(dist(y))
  eps1 <- 1e-10
  # Marker score is combination of specificity and sparsity
  score1 <- sum(d1^spec.exp) / (sum(d1) + eps1)
  return(score1)
}

```


```{r load-data}
root.path <- "//allen/programs/celltypes/workgroups"
all.paths <- paste0(root.path,
                    c("/rnaseqanalysis/shiny/facs_seq/mouse_V1_ALM_20170905_exons_introns/", 
                      "/rnaseqanalysis/shiny/facs_seq/mouse_V1_ALM_20170405/",
                      "/hct/CT_clustering/mouse_cell_vs_nuc/shinydb/20170818_VISp_L5_nuc_1/",
                      "/hct/CT_clustering/mouse_cell_vs_nuc/shinydb/20170818_VISp_L5_cell_1/",
                      "/hct/CT_clustering/mouse_cell_vs_nuc/shinydb/20170818_VISp_L5_snap25_1/",
                      "/hct/CT_clustering/mouse_cell_vs_nuc/shinydb/20170818_VISp_L5_nuc_exon_1/",
                      "/rnaseqanalysis/shiny/facs_seq/mouse_VISp_SMV1_1679/"))
names(all.paths) <- c("Mouse_V1_ALM", "Mouse_V1_ALM_0405", 
                      "Mouse_L5_Nuc", "Mouse_L5_Cell", 
                      "Mouse_L5_Snap25", "Mouse_L5_Nuc_Ex",
                      "Mouse_V1_Tasic")
paths <- all.paths[c("Mouse_L5_Nuc", "Mouse_V1_Tasic")]

dend <- list()
anno <- list()
anno.cl <- list()
prop.all <- list()
for (dat1 in names(paths)) {
  if (file.exists(paste0(paths[[dat1]], "dend.RData"))) {
    dend[[dat1]] <- readRDS(paste0(paths[[dat1]], "dend.RData"))
  }
  anno[[dat1]] <- as.data.frame(read_feather(paste0(paths[[dat1]], "anno.feather")))
  colnames(anno[[dat1]])[colnames(anno[[dat1]]) == "final_label"] <- "cluster_label"  # KLUDGE - Tasic
  anno[[dat1]]$cluster_label <- sub("_outlier", "", anno[[dat1]]$cluster_label)  # KLUDGE - fix MTG labels
  prop1 <- as.data.frame(t(read_feather(paste0(paths[[dat1]], "prop.feather"))))
  rownames(prop1) <- toupper(rownames(prop1))
  
  if (dat1 == "Mouse_V1_ALM") {
    cl.lab <- unique(anno[[dat1]][, c("cluster_id", "cluster_label")])
    cl.lab <- cl.lab[order(cl.lab$cluster_id), ]
    colnames(prop1) <- as.character(cl.lab$cluster_label)
    keep.cl <- unlist(dendrapply(dend[[dat1]], function(x) if (is.leaf(x)) attr(x, "label")))
    keep.cl <- intersect(keep.cl, unique(anno[[dat1]]$cluster_label[! grepl("ALM", anno[[dat1]]$cluster_label)]))
    keep.cl <- intersect(keep.cl, unique(anno[[dat1]]$cluster_label[anno[[dat1]]$category_label %in% c("GABAergic", "Glutamatergic")]))
    anno.cl1 <- unique(anno[[dat1]][, c("cluster_label", "coarse_label", "category_label")])
    confusion.score <- tapply(anno[[dat1]]$confusion_label, anno[[dat1]]$cluster_label, median)
    anno.cl1$confusion_score <- confusion.score[match(anno.cl1$cluster_label, names(confusion.score))]
    
  } else if (dat1 == "Mouse_V1_ALM_0405") {
    colnames(prop1) <- unique(anno[[dat1]]$cluster_label)
    keep.cl <- unlist(dendrapply(dend[[dat1]], function(x) if (is.leaf(x)) attr(x, "label")))
    keep.cl <- intersect(keep.cl, unique(anno[[dat1]]$cluster_label[! grepl("ALM", anno[[dat1]]$cluster_label)]))
    keep.cl <- intersect(keep.cl, unique(anno[[dat1]]$cluster_label[anno[[dat1]]$category_label %in% c("GABAergic", "Glutamatergic")]))
    anno.cl1 <- unique(anno[[dat1]][, c("cluster_label", "coarse_label", "category_label")])
    confusion.score <- tapply(anno[[dat1]]$confusion_label, anno[[dat1]]$cluster_label, median)
    anno.cl1$confusion_score <- confusion.score[match(anno.cl1$cluster_label, names(confusion.score))]
    
  } else if (grepl("Mouse_L5", dat1)) {
    colnames(prop1) <- unique(anno[[dat1]]$cluster_label)
    # keep.cl <- unique(anno[[dat1]]$cluster_label)
    keep.cl <- unlist(dendrapply(dend[[dat1]], function(x) if (is.leaf(x)) attr(x, "label")))
    # keep.cl <- unique(anno[[dat1]][anno[[dat1]]$cluster_type_label %in% c("inh", "exc", "glia"), c("cluster_label")])
    anno.cl1 <- data.frame(cluster_label = unique(anno[[dat1]][, c("cluster_label")]))
  } else if (grepl("Mouse_V1_Tasic", dat1)) {
    anno[[dat1]]$cluster_label <- sub("^[0-9]+ f[0-9]+ ", "", anno[[dat1]]$cluster_label)
    colnames(prop1) <- unique(anno[[dat1]]$cluster_label)
    keep.cl <- unique(anno[[dat1]]$cluster_label[order(anno[[dat1]]$final_id)])
    keep.cl <- keep.cl[keep.cl != ""]
    keep.cl <- keep.cl[c(1:42)]  # Neuronal
    # keep.cl <- keep.cl[c(1:23)]  # Inhibitory
    # keep.cl <- keep.cl[c(1:9,23)]  # L1 Inhibitory
    anno[[dat1]] <- droplevels(subset(anno[[dat1]], cluster_label %in% keep.cl))
    anno.cl1 <- data.frame(cluster_label = unique(anno[[dat1]][, c("cluster_label")]))
  }
  
  # Common annotation
  cl.size <- table(anno[[dat1]]$cluster_label)
  anno.cl1$size <- cl.size[match(names(cl.size), anno.cl1$cluster_label)]
  anno.cl[[dat1]] <- anno.cl1[match(keep.cl, anno.cl1$cluster_label), ]
  anno[[dat1]] <- droplevels(subset(anno[[dat1]], cluster_label %in% keep.cl))
  prop1.subset <- prop1[, match(keep.cl, colnames(prop1))]
  prop.all[[dat1]] <- prop1.subset
}

```


```{r plot-dendrograms}
for (dat1 in names(paths)) {
  pdf(file = paste0("../analysis/match_nuc_cell/", dat1, "_tree.pdf"), 
       width = 4, height = 4)
  plot(dend[[dat1]], ylim = c(0, 1))
  dev.off()
}

```



```{r process-data, echo=FALSE}
prop <- prop.all

shared.genes <- intersect(rownames(prop[[1]]), rownames(prop[[2]]))

cnt <- list()
for (dat1 in names(paths)) {
  prop[[dat1]] <- prop[[dat1]][shared.genes, ]
  cnt[[dat1]] <- apply(prop[[dat1]], 1, function(x) sum(x > 0.5))
}

```


```{r calc-marker-scores}
beta <- list()
for (dat1 in names(paths)) {
  beta[[dat1]] <- apply(prop[[dat1]], 1, calc_beta)
}

```


#### Select genes that mark clusters in both data sets.
```{r subset-genes}
# Empirical trade-off between number of genes and cell type specificity
cnt1.max <- ncol(prop[[1]]) / 2; cnt2.max <- ncol(prop[[2]]) / 2
keep.genes <- which(cnt[[1]] < cnt1.max & cnt[[2]] < cnt2.max & 
                      cnt[[1]] > 0 & cnt[[2]] > 0 & 
                      beta[[1]] > 0.3 & beta[[2]] > 0.3)


prop.subset <- list()
cnt.subset <- list()
score.subset <- list()
for (dat1 in names(paths)) {
  prop.subset[[dat1]] <- prop[[dat1]][keep.genes, ]  # grep("L5a", colnames(prop[[dat1]]))
  cnt.subset[[dat1]] <- cnt[[dat1]][keep.genes]
  score.subset[[dat1]] <- beta[[dat1]][keep.genes]
}

```



```{r match-clusters}
cl.pairs <- expand.grid(colnames(prop.subset[[1]]), colnames(prop.subset[[2]]))

cor.all <- NULL
cor.wt <- score.subset[[1]] * score.subset[[2]]
for (i in 1:nrow(cl.pairs)) {
  cl1 <- cl.pairs[i, 1]
  cl2 <- cl.pairs[i, 2]
  cor1 <- cov.wt(cbind(prop.subset[[1]][, cl1], prop.subset[[2]][, cl2]), wt = cor.wt,
                 center = TRUE, cor = TRUE)$cor[1, 2]  # Weighted cor
  cor.all <- c(cor.all, cor1)
}
names(cor.all) <- apply(cl.pairs, 1, function(x) paste(x, collapse = "~"))

# tail(sort(cor.all), 20)

```


### Compare expression of marker genes between species
#### Correlate proportion of cells/nuclei in each cluster expressing marker genes with FPKM > 1.
```{r plot-cor-heatmap, fig.width = 6.5, fig.height = 4.5}
cor.m <- matrix(0, ncol(prop.subset[[1]]), ncol(prop.subset[[2]]), 
                dimnames = list(colnames(prop.subset[[1]]),
                                colnames(prop.subset[[2]])))

counter <- 0
for (i in 1:nrow(cl.pairs)) {
  counter <- counter + 1
  cl1 <- cl.pairs[i, 1]
  cl2 <- cl.pairs[i, 2]
  cor.m[cl1, cl2] <- cor.all[counter]
}

cor.ml <- melt(cor.m)

g1 <- ggplot(cor.ml, aes(x = Var2, y = Var1, fill = value)) +
  geom_tile(color = "grey80") +
  scale_y_discrete(limits = rev(levels(cor.ml$Var1))) +
  # scale_fill_gradient2(low = "dark blue", high = "red", mid = "white",
  #  midpoint = 0, limit = c(min(cor.m), max(cor.m)), space = "Lab", name="Correlation") +
  guides(fill=guide_colorbar("Correlation")) +
  # scale_fill_gradientn(colors = rev(brewer.pal(n = 7, name = "RdYlBu")),
  #                      guide = "colorbar") +
  scale_fill_gradientn(colors = c(rev(brewer.pal(9, "Blues")), "white",
                                  brewer.pal(9, "Reds")), 
                       guide = "colorbar", limits = c(-1, 1)) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(names(paths)[2]) +
  ylab(names(paths)[1])
plot(g1)
ggsave(g1, filename = paste0("../analysis/match_nuc_cell/", names(paths)[1],
                             "_vs_", names(paths)[2], "_heatmap.pdf"), 
       width = 12, height = 3.5)

```


#### Threshold correlation matrix to identify best matching cell types (light blue), reciprocal best matches (dark blue), and unique best matches (red).
```{r find-cl-homol, fig.width = 20, fig.height = 12}
row.max <- t(apply(cor.m, 1, function(x) ifelse(x == max(x), 1, 0)))
col.max <- apply(cor.m, 2, function(x) ifelse(x == max(x), 2, 0))

cor.m2 <- row.max + col.max
cl.match <- which(cor.m2 == 3)
for (i in cl.match) {
  row.val <- cor.m2[row(cor.m2)[i], ]
  col.val <- cor.m2[, col(cor.m2)[i]]
  if (sum(row.val) == 3 & sum(col.val) == 3) {
    cor.m2[i] <- 4
  }
}


# Scale plot by difference in correlation
cor.diff.row <- t(apply(cor.m, 1, function(x) x - max(x[-which.max(x)])))
cor.diff.col <- apply(cor.m, 2, function(x) x - max(x[-which.max(x)]))
cor.diff.rowl <- melt(cor.diff.row)
cor.diff.coll <- melt(cor.diff.col)
cor.diff <- apply(cbind(cor.diff.rowl$value, cor.diff.coll$value), 1, min)
cor.diffm <- matrix(cor.diff, nrow(cor.m), ncol(cor.m))
cor.m2l <- melt(cor.m2)
cor.m2l$specificity <- ifelse(cor.diff < 0, 0, cor.diff) #- min(cor.diffl$value)
cor.m2l$cor <- cor.ml$value

cor.m2l$homology_type <- as.factor(cor.m2l$value)
homol.df <- data.frame(orig_lev = c("0", "1", "2", "3", "4"), 
                       new_lev = c("None", paste(names(paths)[1], "best match"), 
                                   paste(names(paths)[2], "best match"), 
                                   "Reciprocal best match", "One-to-one"),
                       lev_color = c("#FFFFFF00", "light blue", "pink", "darkorchid1", "darkorchid4"))
levels(cor.m2l$homology_type) <- homol.df$new_lev[match(levels(cor.m2l$homology_type), homol.df$orig_lev)]
homol.pal <- as.character(homol.df$lev_color[match(levels(cor.m2l$homology_type), homol.df$new_lev)])

g2 <- ggplot(cor.m2l, aes(x = Var2, y = Var1, size = specificity, color = homology_type)) +
  geom_point() +
  scale_color_manual(values = homol.pal) +
  theme_bw() +
  scale_y_discrete(limits = rev(levels(cor.m2l$Var1))) +
  scale_size(range = c(3, 8)) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  xlab(names(paths)[2]) +
  ylab(names(paths)[1])
plot(g2)

```

#### For unique best matching clusters, compare cluster proportions
```{r compare-cl-prop, fig.height = 5, fig.width = 5}
cl.match <- cbind(rownames(cor.m2)[apply(cor.m2, 2, which.max)], 
                  colnames(cor.m2))
good.match <- which(apply(cor.m2, 2, max) >= 3 & apply(cor.diffm, 2, max) > 0.0)
cl.to.plot <- cl.match[good.match, ]
cl.to.plot.lab <- sapply(cl.to.plot[, 2], function(x) strsplit(x, "_")[[1]][4])
# cl.to.plot <- cl.match  # Keep all cluster pairs
# cl.to.plot <- cl.to.plot[order(cl.to.plot[, 2]), ]

if (nrow(cl.to.plot) > 0) {
  
  # Cluster sizes
  cl.prop <- list()
  for (i in 1:2) { 
    cl.cnt1 <- as.matrix(table(anno[[i]]$cluster_label))
    cl.cnt.subset <- cl.cnt1[rownames(cl.cnt1) %in% unlist(cl.to.plot), ]
    cl.prop[[i]] <- cl.cnt.subset / sum(cl.cnt.subset)
  }
  
  cl1.prop.subset <- cl.prop[[1]][cl.to.plot[, 1]]
  cl2.prop.subset <- cl.prop[[2]][cl.to.plot[, 2]]
  cl.prop.df <- data.frame(cl.to.plot.lab, cl1.prop.subset, cl2.prop.subset)
  
  g1 <- ggplot(cl.prop.df, aes(x = cl1.prop.subset, y = cl2.prop.subset)) +
    geom_abline(aes(intercept = 0, slope = 1), color = "grey90") +
    geom_point() +
    geom_text_repel(label = cl.to.plot.lab, size = 2) +
    xlim(c(0, 0.25)) +
    ylim(c(0, 0.3)) +
    xlab("Proportion of nuclei in cluster") +
    ylab("Proportion of cells in cluster") +
    theme_bw() +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank())
  plot(g1)
  ggsave(g1, filename = "../analysis/match_nuc_cell/cl_prop_scatter.pdf", 
         width = 3, height = 3)
}

nuc.cell.prop <- data.frame(cluster_pair = cl.to.plot.lab,
                 Nuclei_prop = round(cl1.prop.subset, 2),
                 Cell_prop = round(cl2.prop.subset, 2))
print(nuc.cell.prop[order(nuc.cell.prop$cluster_pair), ])

write.csv(nuc.cell.prop[order(nuc.cell.prop$cluster_pair), ], 
          file = "../analysis/match_nuc_cell/nuc_cell_cl_prop_table.csv",
          row.names = FALSE)
```

```{r compare-cl-confusion-score}
if (all(sapply(anno.cl, function(x) any(names(x) == "confusion_score")))) {
  plot(anno.cl[[1]]$confusion_score[cl.match[, 1]], 
       1 - anno.cl[[2]]$confusion_score[cl.match[, 2]],
       xlab = names(paths)[1], ylab = names(paths)[2])
  text(anno.cl[[1]]$confusion_score[cl.match[good.match, 1]], 
       1 - anno.cl[[2]]$confusion_score[cl.match[good.match, 2]],
       labels = names(good.match))
}
```


#### For unique best matching clusters, plot expression of marker genes.
#### Label genes that are cell type specific in both data sets.
```{r plot-cl-pair, fig.height = 12, fig.width = 12}
if (nrow(cl.to.plot) > 0) {
  shared.marker.genes <- list()
  pdf(file = "../analysis/match_nuc_cell/nuc_cell_cl_match_cor.pdf", 
      width = 8, height = 10)
  par(mfrow = c(ceiling(sqrt(nrow(cl.to.plot))), ceiling(sqrt(nrow(cl.to.plot)))))
  cl1.labs <- paste0("N", 1:11, "_", 
                     sapply(cl.to.plot[, 1], function(x) strsplit(x, "_")[[1]][4]))
  cl2.labs <- paste0("C", 1:11, "_", 
                     sapply(cl.to.plot[, 2], function(x) strsplit(x, "_")[[1]][4]))
  for (i in 1:nrow(cl.to.plot)) {
    cl1 <- cl.to.plot[i, 1]
    cl2 <- cl.to.plot[i, 2]
    cl1.lab <- cl1.labs[i]
    cl2.lab <- cl2.labs[i]
    
    cor1 <- cov.wt(cbind(prop.subset[[1]][, cl1], prop.subset[[2]][, cl2]), wt = cor.wt,
                   center = TRUE, cor = TRUE)$cor[1, 2]  # Weighted cor
    label.genes <- prop.subset[[1]][, cl1] > 0.5 & prop.subset[[2]][, cl2] > 0.5 &
      cnt.subset[[1]] == 1 & cnt.subset[[2]] == 1

    # label.genes <- prop.subset[[1]][, cl1] > 0.5 & prop.subset[[2]][, cl2] < 0.1 |
    #   prop.subset[[1]][, cl1] < 0.1 & prop.subset[[2]][, cl2] > 0.5
    
    plot(jitter(prop.subset[[1]][, cl1]),
         jitter(prop.subset[[2]][, cl2]), 
         pch = 19, cex = 0.2, col = "grey", las = 1,
         xlim = c(0, 1), ylim = c(0, 1), xlab = cl1.lab, ylab = cl2.lab, 
         main = paste("r =", signif(cor1, 2)))
    abline(0, 1, col = "grey")
    if (sum(label.genes) > 0) {
      text(jitter(prop.subset[[1]][label.genes, cl1]),
           jitter(prop.subset[[2]][label.genes, cl2]), 
           rownames(prop.subset[[1]])[label.genes], cex = 0.6)
      cl.pair.name <- paste(cl1.lab, cl2.lab)
      shared.marker.genes[[cl.pair.name]] <- paste0(rownames(prop.subset[[1]])[label.genes],
                                                    collapse = ",")
    }
  }
  dev.off()
}

```
